<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Workshop on Tools for Reproducible Research"><meta name=author content="Jacques Dainat et al."><link href=https://southgreenplatform.github.io/training_reproducible_research/pages/rmarkdown/r-markdown-7-the-mrsa-case-study/ rel=canonical><link href=../r-markdown-6-rendering/ rel=prev><link href=../r-markdown-8-extra-material/ rel=next><link rel=icon href=../../../assets/images/favicon.png><meta name=generator content="mkdocs-1.4.3, mkdocs-material-9.1.15"><title>The MRSA case study - Bioinformatics Workshop on Tools for Reproducible Research</title><link rel=stylesheet href=../../../assets/stylesheets/main.26e3688c.min.css><link rel=stylesheet href=../../../assets/stylesheets/palette.ecc896b0.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../../stylesheets/extra.css><script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#overview class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-header__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../images/southgreen.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bioinformatics Workshop on Tools for Reproducible Research </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> The MRSA case study </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../../introduction/introduction_southgreen/ class="md-tabs__link md-tabs__link--active"> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../cheat_sheet/bash/bash/ class=md-tabs__link> Cheat sheet </a> </li> <li class=md-tabs__item> <a href=../../slido/ class=md-tabs__link> Ask a question </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-nav__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../images/southgreen.png alt=logo> </a> Bioinformatics Workshop on Tools for Reproducible Research </label> <div class=md-nav__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../.. class=md-nav__link> Home </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Tutorials <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../introduction/introduction_southgreen/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../introduction/roundtable/ class=md-nav__link> Roundtable Introduce Yourself </a> </li> <li class=md-nav__item> <a href=../../introduction/introduction_rr/ class=md-nav__link> Introduction Reproducible Research </a> </li> <li class=md-nav__item> <a href=../../data-management/data-management/ class=md-nav__link> Data Management </a> </li> <li class=md-nav__item> <a href=../../ice-breaker/ class=md-nav__link> Ice breaker session </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_6> <label class=md-nav__link for=__nav_2_6 id=__nav_2_6_label tabindex=0> Git <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_6_label aria-expanded=false> <label class=md-nav__title for=__nav_2_6> <span class="md-nav__icon md-icon"></span> Git </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../git/git-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../git/git-setup/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../git/git-2-creating-repositories/ class=md-nav__link> Creating a git repository </a> </li> <li class=md-nav__item> <a href=../../git/git-3-committing-changes/ class=md-nav__link> Committing changes </a> </li> <li class=md-nav__item> <a href=../../git/git-undoing-and-recovering/ class=md-nav__link> Undoing and recovering </a> </li> <li class=md-nav__item> <a href=../../git/git-4-ignoring-files/ class=md-nav__link> Ignoring files </a> </li> <li class=md-nav__item> <a href=../../git/git-5-branches/ class=md-nav__link> Branching and merging </a> </li> <li class=md-nav__item> <a href=../../git/git-6-tags/ class=md-nav__link> Tagging commits </a> </li> <li class=md-nav__item> <a href=../../git/git-7-working-remotely/ class=md-nav__link> Working remotly </a> </li> <li class=md-nav__item> <a href=../../git/git-8-conflicts/ class=md-nav__link> Conflict </a> </li> <li class=md-nav__item> <a href=../../git/git-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_7> <label class=md-nav__link for=__nav_2_7 id=__nav_2_7_label tabindex=0> Conda <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_7_label aria-expanded=false> <label class=md-nav__title for=__nav_2_7> <span class="md-nav__icon md-icon"></span> Conda </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../conda/conda-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../conda/conda-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../conda/conda-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../conda/conda-3-projects/ class=md-nav__link> Projects </a> </li> <li class=md-nav__item> <a href=../../conda/conda-4-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8> <label class=md-nav__link for=__nav_2_8 id=__nav_2_8_label tabindex=0> Workflow Manager System <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_8_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Workflow Manager System </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../wms/ class=md-nav__link> Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_2> <label class=md-nav__link for=__nav_2_8_2 id=__nav_2_8_2_label tabindex=0> Snakemake <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8_2> <span class="md-nav__icon md-icon"></span> Snakemake </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../snakemake/snakemake-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-3-visualising-workflows/ class=md-nav__link> Visualising workflow </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-4-the-mrsa-workflow/ class=md-nav__link> The MRSA Workflow </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-5-parameters/ class=md-nav__link> Parameters </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-6-logs/ class=md-nav__link> Logs </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-7-temporary-files/ class=md-nav__link> Temporary files </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-8-targets/ class=md-nav__link> Targets </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-9-shadow-rules/ class=md-nav__link> Shadow rules </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-10-generalizing-workflows/ class=md-nav__link> Generalizing workflows </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-11-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_3> <label class=md-nav__link for=__nav_2_8_3 id=__nav_2_8_3_label tabindex=0> Nextflow <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_3_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8_3> <span class="md-nav__icon md-icon"></span> Nextflow </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../nextflow/nextflow-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-nfcore/ class=md-nav__link> Getting started with nf-core </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-3-executing-workflows/ class=md-nav__link> Executing workflows </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-4-working-with-processes/ class=md-nav__link> Working with process </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-5-workflow-configuration/ class=md-nav__link> Workflow configuration </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-6-optimising-the-mrsa-workflow/ class=md-nav__link> Optmising the MRSA workflow </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-7-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_9> <label class=md-nav__link for=__nav_2_9 id=__nav_2_9_label tabindex=0> Containers <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_9_label aria-expanded=false> <label class=md-nav__title for=__nav_2_9> <span class="md-nav__icon md-icon"></span> Containers </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../containers/containers-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../containers/containers-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../containers/containers-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../containers/containers-3-building-images/ class=md-nav__link> Building a Docker image </a> </li> <li class=md-nav__item> <a href=../../containers/containers-4-managing-containers/ class=md-nav__link> Managing containers </a> </li> <li class=md-nav__item> <a href=../../containers/containers-5-sharing-images/ class=md-nav__link> Distributing your images </a> </li> <li class=md-nav__item> <a href=../../containers/containers-6-packaging-the-case-study/ class=md-nav__link> Packaging the case study </a> </li> <li class=md-nav__item> <a href=../../containers/containers-7-singularity/ class=md-nav__link> Singularity </a> </li> <li class=md-nav__item> <a href=../../containers/containers-8-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_10> <label class=md-nav__link for=__nav_2_10 id=__nav_2_10_label tabindex=0> Notebook <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_10_label aria-expanded=false> <label class=md-nav__title for=__nav_2_10> <span class="md-nav__icon md-icon"></span> Notebook </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../notebook/notebook-context/ class=md-nav__link> Context </a> </li> <li class=md-nav__item> <a href=../../notebook/notebook-markdown/ class=md-nav__link> Markdown </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_11 checked> <label class=md-nav__link for=__nav_2_11 id=__nav_2_11_label tabindex=0> R Markdown <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_11_label aria-expanded=true> <label class=md-nav__title for=__nav_2_11> <span class="md-nav__icon md-icon"></span> R Markdown </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../r-markdown-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../r-markdown-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../r-markdown-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../r-markdown-3-text/ class=md-nav__link> Text </a> </li> <li class=md-nav__item> <a href=../r-markdown-4-code-chunks/ class=md-nav__link> Code chunks </a> </li> <li class=md-nav__item> <a href=../r-markdown-5-the-yaml-header/ class=md-nav__link> The YAML header </a> </li> <li class=md-nav__item> <a href=../r-markdown-6-rendering/ class=md-nav__link> Rendering </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> The MRSA case study <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> The MRSA case study </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#overview class=md-nav__link> Overview </a> </li> <li class=md-nav__item> <a href=#rendering-options-and-paths class=md-nav__link> Rendering options and paths </a> </li> <li class=md-nav__item> <a href=#formatting-tables-and-figures class=md-nav__link> Formatting tables and figures </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../r-markdown-8-extra-material/ class=md-nav__link> Extra materiel </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_12> <label class=md-nav__link for=__nav_2_12 id=__nav_2_12_label tabindex=0> Jupyter <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_12_label aria-expanded=false> <label class=md-nav__title for=__nav_2_12> <span class="md-nav__icon md-icon"></span> Jupyter </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../jupyter/jupyter-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-3-magics/ class=md-nav__link> Magics </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-4-plotting/ class=md-nav__link> Plotting </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-5-widgets/ class=md-nav__link> Widgets </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-6-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-7-converting-notebooks/ class=md-nav__link> Converting notebooks </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-8-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_13> <label class=md-nav__link for=__nav_2_13 id=__nav_2_13_label tabindex=0> Quarto <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_13_label aria-expanded=false> <label class=md-nav__title for=__nav_2_13> <span class="md-nav__icon md-icon"></span> Quarto </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../quarto/quarto-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-3-the-rendering/ class=md-nav__link> Rendering </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-4-the-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-5-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-6-comparison/ class=md-nav__link> Comparison </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-7-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_14> <label class=md-nav__link for=__nav_2_14 id=__nav_2_14_label tabindex=0> Making a project reproducible <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_14_label aria-expanded=false> <label class=md-nav__title for=__nav_2_14> <span class="md-nav__icon md-icon"></span> Making a project reproducible </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../project/project-1-review/ class=md-nav__link> Review </a> </li> <li class=md-nav__item> <a href=../../project/project-2-good_praticies/ class=md-nav__link> Good praticies </a> </li> <li class=md-nav__item> <a href=../../project/project-3-and_now/ class=md-nav__link> And now ? </a> </li> <li class=md-nav__item> <a href=../../project/project-4-demo/ class=md-nav__link> To end on a high note ! </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../take_down/take_down/ class=md-nav__link> Take down </a> </li> <li class=md-nav__item> <a href=../../final_word/ class=md-nav__link> Final word </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> Cheat sheet <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Cheat sheet </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../cheat_sheet/bash/bash/ class=md-nav__link> Bash </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/git/git/ class=md-nav__link> Git </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/git/git_official/ class=md-nav__link> Git official </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/conda/conda/ class=md-nav__link> Conda </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/notebook/notebook/ class=md-nav__link> Notebook </a> </li> <li class=md-nav__item> <a href=https://dockerlabs.collabnix.com/docker/cheatsheet/ class=md-nav__link> Docker </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../slido/ class=md-nav__link> Ask a question </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/edit/main/docs/pages/rmarkdown/r-markdown-7-the-mrsa-case-study.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg> </a> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/raw/main/docs/pages/rmarkdown/r-markdown-7-the-mrsa-case-study.md title="View source of this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.15 8.15 0 0 1-1.23-2Z"/></svg> </a> <h1>The MRSA case study</h1> <p>As you might remember from the <a href=introduction>intro</a>, we are attempting to understand how lytic bacteriophages can be used as a future therapy for the multiresistant bacteria MRSA (methicillin-resistant <em>Staphylococcus aureus</em>). In this exercise, we will use R Markdown to make a report in form of a Supplementary Material HTML based on the outputs from the <a href=snakemake-10-generalizing-workflows>Snakemake tutorial</a>. Among the benefits of having the supplementary material (or even the full manuscript) in R Markdown format are:</p> <ul> <li>It is fully transparent how the text, tables and figures were produced.</li> <li>If you get reviewer comments, or realize you've made a mistake somewhere, you can easily update your code and regenerate the document with the push of a button.</li> <li>By making report generation part of your workflow early on in a project, much of the eventual manuscript "writes itself". You no longer first have to finish the research part and <em>then</em> start creating the tables and figures for the paper.</li> </ul> <p>Before you start:</p> <ul> <li>Make sure that your working directory in R is <code>training-reproducible-research-area/rmarkdown</code> in the course directory (Session &gt; Set Working Directory).</li> <li>Open the file <code>code/supplementary_material.Rmd</code>.</li> </ul> <div class="admonition note"> <p class=admonition-title>Note</p> <p>In this tutorial we have used Conda to install all the R packages we need, so that you get to practice how you can actually do this in projects of your own. You can, however, install things using <code>install.packages()</code> or <code>BiocManager::install()</code> as well, even though this makes it both less reproducible and more complicated in most cases.</p> </div> <h2 id=overview>Overview<a class=headerlink href=#overview title="Permanent link">#</a></h2> <p>Let's start by taking a look at the YAML header at the top of the file. The parameters correspond to files (and sample IDs) that are generated by the MRSA analysis workflow (see the <a href=../../snakemake/snakemake-10-generalizing-workflows>Snakemake tutorial</a>) and contain results that we want to include in the supplementary material document. We've also specified that we want to render to HTML.</p> <div class=highlight><pre><span></span><code><span class=nn>---</span>
<span class=nt>title</span><span class=p>:</span><span class=w> </span><span class=s>&quot;Supplementary</span><span class=nv> </span><span class=s>Materials&quot;</span>
<span class=nt>output</span><span class=p>:</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">html_document</span>
<span class=nt>params</span><span class=p>:</span>
<span class=w>    </span><span class=nt>counts_file</span><span class=p>:</span><span class=w> </span><span class=s>&quot;results/tables/counts.tsv&quot;</span>
<span class=w>    </span><span class=nt>multiqc_file</span><span class=p>:</span><span class=w> </span><span class=s>&quot;intermediate/multiqc_general_stats.txt&quot;</span>
<span class=w>    </span><span class=nt>summary_file</span><span class=p>:</span><span class=w> </span><span class=s>&quot;results/tables/counts.tsv.summary&quot;</span>
<span class=w>    </span><span class=nt>rulegraph_file</span><span class=p>:</span><span class=w> </span><span class=s>&quot;results/rulegraph.png&quot;</span>
<span class=w>    </span><span class=nt>SRR_IDs</span><span class=p>:</span><span class=w> </span><span class=s>&quot;SRR935090</span><span class=nv> </span><span class=s>SRR935091</span><span class=nv> </span><span class=s>SRR935092&quot;</span>
<span class=w>    </span><span class=nt>GSM_IDs</span><span class=p>:</span><span class=w> </span><span class=s>&quot;GSM1186459</span><span class=nv> </span><span class=s>GSM1186460</span><span class=nv> </span><span class=s>GSM1186461&quot;</span>
<span class=nn>---</span>
</code></pre></div> <ul> <li>From a reproducibility perspective it definitely makes sense to include information about who authored the document and the date it was generated. Add the two lines below to the YAML header. Note that we can include inline R code by using <code>`r some_code`</code>.</li> </ul> <div class=highlight><pre><span></span><code>author: John Doe, Joan Dough, Jan Doh, Dyon Do
date: &quot;`r format(Sys.time(), &#39;%d %B, %Y&#39;)`&quot;
</code></pre></div> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>Make it a practice to keep track of all input files and add them as parameters rather than hard-coding them later in the R code.</p> </div> <p>Next, take a look at the <code>dependencies</code>, <code>read_params</code>, and <code>read_data</code> chunks. They </p> <ol> <li>load the required packages, </li> <li>read the parameters and store them in R objects to be used later in the code, and </li> <li>read the data in the counts file, the multiqc file, as well as fetch meta data from GEO. These chunks are provided as is, and you do not need to edit them.</li> </ol> <p>Below these chunks there is some markdown text that contains the Supplementary Methods section. Note the use of section headers using <code>#</code> and <code>##</code>. Then there is a Supplementary Tables and Figures section. This contains four code chunks, each for a specific table or figure. Have a quick look at the code and see if you can figure out what it does, but don't worry if you can't understand everything.</p> <p>Finally, there is a <em>Reproducibility</em> section which describes how the results in the report can be reproduced. The <code>session_info</code> chunk prints information regarding R version and which packages and versions that are used. We highly encourage you to include this chunk in all your R Markdown reports: it's an effortless way to increase reproducibility.</p> <h2 id=rendering-options-and-paths>Rendering options and paths<a class=headerlink href=#rendering-options-and-paths title="Permanent link">#</a></h2> <p>Now that you have had a look at the R Markdown document, it is time to Knit! We will do this from the R terminal (rather than pressing <em>Knit</em>).</p> <div class=highlight><pre><span></span><code><span class=n>rmarkdown</span><span class=o>::</span><span class=nf>render</span><span class=p>(</span><span class=s>&quot;code/supplementary_material.Rmd&quot;</span><span class=p>,</span><span class=w> </span><span class=n>output_dir</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;results&quot;</span><span class=p>)</span>
</code></pre></div> <p>The reason for this is that we can then redirect the output html file to be saved in the <code>results/</code> directory.</p> <p>Normally, while rendering, R code in the Rmd file will be executed using the directory of the Rmd file as working directory (<code>rmarkdown/code</code> in this case). However, it is good practice to write all code as if it would be executed from the project root directory (<code>rmarkdown/</code> in this case). For instance, you can see that we have specified the files in <code>params</code> with relative paths from the project root directory. To set a different directory as working directory for all chunks one modifies the knit options like this:</p> <div class=highlight><pre><span></span><code><span class=n>knitr</span><span class=o>::</span><span class=n>opts_knit</span><span class=o>$</span><span class=nf>set</span><span class=p>(</span><span class=n>root.dir</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&#39;../&#39;</span><span class=p>)</span>
</code></pre></div> <p>Here we set the working directory to the parent directory of the Rmd file (<code>../</code>), in other words, the project root. Use this rather than <code>setwd()</code> while working with Rmd files.</p> <ul> <li>Take a look at the output. You should find the html file in the <code>results</code> directory.</li> </ul> <h2 id=formatting-tables-and-figures>Formatting tables and figures<a class=headerlink href=#formatting-tables-and-figures title="Permanent link">#</a></h2> <p>You will probably get a good idea of the contents of the file, but the tables look weird and the figures could be better formatted. Let's start by adjusting the figures!</p> <ul> <li> <p>Locate the <code>Setup</code> chunk. Here, we have already set <code>echo = FALSE</code>. Let's add some default figure options: <code>fig.height = 6, fig.width = 6, fig.align = 'center'</code>. This will make the figures slightly smaller than default and center them.</p> </li> <li> <p>Knit again, using the same R command as above. Do you notice any difference? Better, but still not perfect!</p> </li> </ul> <p>Let's improve the tables! We have not talked about tables before. There are several options to print tables, here we will use the <code>kable</code> function which is part of the <code>knitr</code> package.</p> <ul> <li>Go to the <code>Sample info</code> chunk. Replace the last line, <code>sample_info</code>, with:</li> </ul> <div class=highlight><pre><span></span><code><span class=n>knitr</span><span class=o>::</span><span class=nf>kable</span><span class=p>(</span><span class=n>sample_info</span><span class=p>)</span>
</code></pre></div> <ul> <li>Knit again and look at the result. You should see a formatted table.</li> <li>The column names can be improved, and we could use a table legend. Change to use the following:</li> </ul> <div class=highlight><pre><span></span><code><span class=n>knitr</span><span class=o>::</span><span class=nf>kable</span><span class=p>(</span><span class=n>sample_info</span><span class=p>,</span><span class=w> </span><span class=n>caption</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;Sample info&quot;</span><span class=p>,</span>
<span class=w>             </span><span class=n>col.names</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>c</span><span class=p>(</span><span class=s>&quot;SRR&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;GEO&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;Strain&quot;</span><span class=p>,</span><span class=w> </span><span class=s>&quot;Treatment&quot;</span><span class=p>))</span>
</code></pre></div> <ul> <li>Knit and check the result.</li> <li>Try to fix the table in the <code>QC statistics</code> chunk in the same manner. The column names are fine here so no need to change them, but add a table legend: "QC stats from FastQC". Knit and check your results.</li> </ul> <p>Let's move on to the figures!</p> <ul> <li>Go to the <code>Counts barplot</code> chunk. To add a figure legend we have to use a chunk option (so not in the same way as for tables). Add the chunk option:</li> </ul> <div class=highlight><pre><span></span><code><span class=n>fig.cap</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;Counting statistics per sample, in terms of read counts for genes</span>
<span class=s>           and reads not counted for various reasons.&quot;</span>
</code></pre></div> <ul> <li>Knit and check the outcome!</li> <li>Next, add a figure legend to the figure in the <code>gene-heatmap</code> chunk. Here we can try out the possibility to add R code to generate the legend:</li> </ul> <div class=highlight><pre><span></span><code><span class=n>fig.cap</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=nf>paste0</span><span class=p>(</span><span class=s>&quot;Expression (log-10 counts) of genes with at least &quot;</span><span class=p>,</span>
<span class=w>                 </span><span class=n>max_cutoff</span><span class=p>,</span><span class=w> </span><span class=s>&quot; counts in one sample and a CV&gt;&quot;</span><span class=p>,</span><span class=w> </span><span class=n>cv_cutoff</span><span class=p>,</span><span class=w> </span><span class=s>&quot;.&quot;</span><span class=p>)</span>
</code></pre></div> <p>This will use the <code>cv_cutoff</code> and <code>max_cutoff</code> variables to ensure that the figure legend gives the same information as was used to generate the plot. Note that figure legends are generated <em>after</em> the corresponding code chunk is evaluated. This means we can use objects defined in the code chunk in the legend.</p> <ul> <li>Knit and have a look at the results.</li> </ul> <p>The heatmap still looks a bit odd. Let's play with the <code>fig.height</code> and <code>out.height</code> options, like we did above, to scale the figure in a more appropriate way. Add this to the chunk options: <code>fig.height = 10, out.height = "22cm"</code>. Knit and check the results. Does it look better now?</p> <ul> <li>Now let's add a third figure! This time we will not plot a figure in R, but use an available image file showing the structure of the Snakemake workflow used to generate the inputs for this report. Add a new chunk at the end of the Supplementary Tables and Figures section containing this code:</li> </ul> <div class=highlight><pre><span></span><code><span class=n>knitr</span><span class=o>::</span><span class=nf>include_graphics</span><span class=p>(</span><span class=nf>normalizePath</span><span class=p>(</span><span class=n>rulegraph_file</span><span class=p>))</span>
</code></pre></div> <blockquote> <p><strong>Knitr paths</strong> <br> Just like for R Markdown paths in general, Knitr needs everything to be relative to the directory in which the <code>.Rmd</code> file is situated. Just like setting the <code>root.dir</code> chunk option can help with this for the overall rendering, the <code>normalizePath()</code> function is needed when using <code>include_graphics()</code> for adding images with Knitr.</p> </blockquote> <ul> <li>Also, add the chunk options:</li> </ul> <div class=highlight><pre><span></span><code><span class=n>fig.cap</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;A rule graph showing the different steps of the bioinformatic</span>
<span class=s>           analysis that is included in the Snakemake workflow.&quot;</span>
</code></pre></div> <p>and:</p> <div class=highlight><pre><span></span><code><span class=n>out.height</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;11cm&quot;</span>
</code></pre></div> <ul> <li>Knit and check the results.</li> </ul> <div class="admonition note"> <p class=admonition-title>Note</p> <p>It is definitely possible to render R Markdown documents as part of a Snakemake or Nextflow workflow. This is something we do for the final version of the MRSA project (in the Containers tutorial). In such cases it is advisable to manage the installation of R and required R packages through your conda environment file and use the <code>rmarkdown::render()</code> command from the shell section of your Snakemake rule or Nexflow process.</p> </div> <div class="admonition success"> <p class=admonition-title>Quick recap</p> <p>In this section you learned some additional details for making nice R Markdown reports in a reproducible research project setting, including setting the root directory, adding tables as well as setting figure and table captions.</p> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../r-markdown-6-rendering/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Rendering" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Rendering </div> </div> </a> <a href=../r-markdown-8-extra-material/ class="md-footer__link md-footer__link--next" aria-label="Next: Extra materiel" rel=next> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Extra materiel </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../../..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "header.autohide", "navigation.footer", "navigation.indexes", "navigation.tabs", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "toc.integrate"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../../../assets/javascripts/bundle.b4d07000.min.js></script> </body> </html>