<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Workshop on Tools for Reproducible Research"><meta name=author content="Jacques Dainat et al."><link href=https://southgreenplatform.github.io/training_reproducible_research/pages/snakemake/snakemake-10-generalizing-workflows/ rel=canonical><link href=../snakemake-9-shadow-rules/ rel=prev><link href=../snakemake-11-extra-material/ rel=next><link rel=icon href=../../../assets/images/favicon.png><meta name=generator content="mkdocs-1.4.3, mkdocs-material-9.1.15"><title>Generalizing workflows - Bioinformatics Workshop on Tools for Reproducible Research</title><link rel=stylesheet href=../../../assets/stylesheets/main.26e3688c.min.css><link rel=stylesheet href=../../../assets/stylesheets/palette.ecc896b0.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../../stylesheets/extra.css><script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-header__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../images/southgreen.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bioinformatics Workshop on Tools for Reproducible Research </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Generalizing workflows </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../../introduction/introduction_southgreen/ class="md-tabs__link md-tabs__link--active"> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../cheat_sheet/bash/bash/ class=md-tabs__link> Cheat sheet </a> </li> <li class=md-tabs__item> <a href=../../slido/ class=md-tabs__link> Ask a question </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-nav__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../images/southgreen.png alt=logo> </a> Bioinformatics Workshop on Tools for Reproducible Research </label> <div class=md-nav__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../.. class=md-nav__link> Home </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Tutorials <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../introduction/introduction_southgreen/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../introduction/roundtable/ class=md-nav__link> Roundtable Introduce Yourself </a> </li> <li class=md-nav__item> <a href=../../introduction/introduction_rr/ class=md-nav__link> Introduction Reproducible Research </a> </li> <li class=md-nav__item> <a href=../../data-management/data-management/ class=md-nav__link> Data Management </a> </li> <li class=md-nav__item> <a href=../../ice-breaker/ class=md-nav__link> Ice breaker session </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_6> <label class=md-nav__link for=__nav_2_6 id=__nav_2_6_label tabindex=0> Git <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_6_label aria-expanded=false> <label class=md-nav__title for=__nav_2_6> <span class="md-nav__icon md-icon"></span> Git </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../git/git-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../git/git-setup/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../git/git-2-creating-repositories/ class=md-nav__link> Creating a git repository </a> </li> <li class=md-nav__item> <a href=../../git/git-3-committing-changes/ class=md-nav__link> Committing changes </a> </li> <li class=md-nav__item> <a href=../../git/git-undoing-and-recovering/ class=md-nav__link> Undoing and recovering </a> </li> <li class=md-nav__item> <a href=../../git/git-4-ignoring-files/ class=md-nav__link> Ignoring files </a> </li> <li class=md-nav__item> <a href=../../git/git-5-branches/ class=md-nav__link> Branching and merging </a> </li> <li class=md-nav__item> <a href=../../git/git-6-tags/ class=md-nav__link> Tagging commits </a> </li> <li class=md-nav__item> <a href=../../git/git-7-working-remotely/ class=md-nav__link> Working remotly </a> </li> <li class=md-nav__item> <a href=../../git/git-8-conflicts/ class=md-nav__link> Conflict </a> </li> <li class=md-nav__item> <a href=../../git/git-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_7> <label class=md-nav__link for=__nav_2_7 id=__nav_2_7_label tabindex=0> Conda <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_7_label aria-expanded=false> <label class=md-nav__title for=__nav_2_7> <span class="md-nav__icon md-icon"></span> Conda </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../conda/conda-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../conda/conda-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../conda/conda-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../conda/conda-3-projects/ class=md-nav__link> Projects </a> </li> <li class=md-nav__item> <a href=../../conda/conda-4-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8 checked> <label class=md-nav__link for=__nav_2_8 id=__nav_2_8_label tabindex=0> Workflow Manager System <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_8_label aria-expanded=true> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Workflow Manager System </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../wms/ class=md-nav__link> Introduction </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_2 checked> <label class=md-nav__link for=__nav_2_8_2 id=__nav_2_8_2_label tabindex=0> Snakemake <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2_8_2> <span class="md-nav__icon md-icon"></span> Snakemake </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../snakemake-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../snakemake-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../snakemake-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../snakemake-3-visualising-workflows/ class=md-nav__link> Visualising workflow </a> </li> <li class=md-nav__item> <a href=../snakemake-4-the-mrsa-workflow/ class=md-nav__link> The MRSA Workflow </a> </li> <li class=md-nav__item> <a href=../snakemake-5-parameters/ class=md-nav__link> Parameters </a> </li> <li class=md-nav__item> <a href=../snakemake-6-logs/ class=md-nav__link> Logs </a> </li> <li class=md-nav__item> <a href=../snakemake-7-temporary-files/ class=md-nav__link> Temporary files </a> </li> <li class=md-nav__item> <a href=../snakemake-8-targets/ class=md-nav__link> Targets </a> </li> <li class=md-nav__item> <a href=../snakemake-9-shadow-rules/ class=md-nav__link> Shadow rules </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <a href=./ class="md-nav__link md-nav__link--active"> Generalizing workflows </a> </li> <li class=md-nav__item> <a href=../snakemake-11-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_3> <label class=md-nav__link for=__nav_2_8_3 id=__nav_2_8_3_label tabindex=0> Nextflow <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_3_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8_3> <span class="md-nav__icon md-icon"></span> Nextflow </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../nextflow/nextflow-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-nfcore/ class=md-nav__link> Getting started with nf-core </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-3-executing-workflows/ class=md-nav__link> Executing workflows </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-4-working-with-processes/ class=md-nav__link> Working with process </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-5-workflow-configuration/ class=md-nav__link> Workflow configuration </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-6-optimising-the-mrsa-workflow/ class=md-nav__link> Optmising the MRSA workflow </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-7-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_9> <label class=md-nav__link for=__nav_2_9 id=__nav_2_9_label tabindex=0> Containers <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_9_label aria-expanded=false> <label class=md-nav__title for=__nav_2_9> <span class="md-nav__icon md-icon"></span> Containers </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../containers/containers-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../containers/containers-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../containers/containers-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../containers/containers-3-building-images/ class=md-nav__link> Building a Docker image </a> </li> <li class=md-nav__item> <a href=../../containers/containers-4-managing-containers/ class=md-nav__link> Managing containers </a> </li> <li class=md-nav__item> <a href=../../containers/containers-5-sharing-images/ class=md-nav__link> Distributing your images </a> </li> <li class=md-nav__item> <a href=../../containers/containers-6-packaging-the-case-study/ class=md-nav__link> Packaging the case study </a> </li> <li class=md-nav__item> <a href=../../containers/containers-7-singularity/ class=md-nav__link> Singularity </a> </li> <li class=md-nav__item> <a href=../../containers/containers-8-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_10> <label class=md-nav__link for=__nav_2_10 id=__nav_2_10_label tabindex=0> Notebook <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_10_label aria-expanded=false> <label class=md-nav__title for=__nav_2_10> <span class="md-nav__icon md-icon"></span> Notebook </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../notebook/notebook-context/ class=md-nav__link> Context </a> </li> <li class=md-nav__item> <a href=../../notebook/notebook-markdown/ class=md-nav__link> Markdown </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_11> <label class=md-nav__link for=__nav_2_11 id=__nav_2_11_label tabindex=0> R Markdown <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_11_label aria-expanded=false> <label class=md-nav__title for=__nav_2_11> <span class="md-nav__icon md-icon"></span> R Markdown </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-3-text/ class=md-nav__link> Text </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-4-code-chunks/ class=md-nav__link> Code chunks </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-5-the-yaml-header/ class=md-nav__link> The YAML header </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-6-rendering/ class=md-nav__link> Rendering </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-7-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-8-extra-material/ class=md-nav__link> Extra materiel </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_12> <label class=md-nav__link for=__nav_2_12 id=__nav_2_12_label tabindex=0> Jupyter <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_12_label aria-expanded=false> <label class=md-nav__title for=__nav_2_12> <span class="md-nav__icon md-icon"></span> Jupyter </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../jupyter/jupyter-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-3-magics/ class=md-nav__link> Magics </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-4-plotting/ class=md-nav__link> Plotting </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-5-widgets/ class=md-nav__link> Widgets </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-6-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-7-converting-notebooks/ class=md-nav__link> Converting notebooks </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-8-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_13> <label class=md-nav__link for=__nav_2_13 id=__nav_2_13_label tabindex=0> Quarto <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_13_label aria-expanded=false> <label class=md-nav__title for=__nav_2_13> <span class="md-nav__icon md-icon"></span> Quarto </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../quarto/quarto-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-3-the-rendering/ class=md-nav__link> Rendering </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-4-the-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-5-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-6-comparison/ class=md-nav__link> Comparison </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-7-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_14> <label class=md-nav__link for=__nav_2_14 id=__nav_2_14_label tabindex=0> Making a project reproducible <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_14_label aria-expanded=false> <label class=md-nav__title for=__nav_2_14> <span class="md-nav__icon md-icon"></span> Making a project reproducible </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../project/project-1-review/ class=md-nav__link> Review </a> </li> <li class=md-nav__item> <a href=../../project/project-2-good_praticies/ class=md-nav__link> Good praticies </a> </li> <li class=md-nav__item> <a href=../../project/project-3-and_now/ class=md-nav__link> And now ? </a> </li> <li class=md-nav__item> <a href=../../project/project-4-demo/ class=md-nav__link> To end on a high note ! </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../take_down/take_down/ class=md-nav__link> Take down </a> </li> <li class=md-nav__item> <a href=../../final_word/ class=md-nav__link> Final word </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> Cheat sheet <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Cheat sheet </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../cheat_sheet/bash/bash/ class=md-nav__link> Bash </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/git/git/ class=md-nav__link> Git </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/git/git_official/ class=md-nav__link> Git official </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/conda/conda/ class=md-nav__link> Conda </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/notebook/notebook/ class=md-nav__link> Notebook </a> </li> <li class=md-nav__item> <a href=https://dockerlabs.collabnix.com/docker/cheatsheet/ class=md-nav__link> Docker </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../slido/ class=md-nav__link> Ask a question </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/edit/main/docs/pages/snakemake/snakemake-10-generalizing-workflows.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg> </a> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/raw/main/docs/pages/snakemake/snakemake-10-generalizing-workflows.md title="View source of this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.15 8.15 0 0 1-1.23-2Z"/></svg> </a> <h1>Generalizing workflows</h1> <p>It's generally a good idea to separate project-specific parameters from the actual implementation of the workflow. If we want to move all project-specific information to <code>config.yml</code>, and let the Snakefile be a more general RNA-seq analysis workflow, we need the config file to:</p> <ul> <li>Specify which samples to run.</li> <li>Specify which genome to align to and where to download its sequence and annotation files.</li> <li>(Any other parameters we might need to make it into a general workflow, <em>e.g.</em> to support both paired-end and single-read sequencing)</li> </ul> <div class="admonition note"> <p class=admonition-title>Note</p> <p>Putting all configuration in <code>config.yml</code> will break the <code>generate_rulegraph</code> rule. You can fix it either by replacing <code>--config max_reads=0</code> with <code>--configfile=config.yml</code> in the shell command of that rule in the Snakefile, or by adding <code>configfile: "config.yml"</code> to the top of the Snakefile (as mentioned in a previous tip).</p> </div> <p>The first point is straightforward; rather than using <code>SAMPLES = ["..."]</code> in the Snakefile we define it as a parameter in <code>config.yml</code>. You can either add it as a list similar to the way it was expressed before by adding <code>SAMPLES: ["..."]</code> to <code>config.yml</code>, or you can use this yaml notation:</p> <div class=highlight><pre><span></span><code><span class=nt>sample_ids</span><span class=p>:</span>
<span class="p p-Indicator">-</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">SRR935090</span>
<span class="p p-Indicator">-</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">SRR935091</span>
<span class="p p-Indicator">-</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">SRR935092</span>
</code></pre></div> <p>You also have to change the workflow to reference <code>config["sample_ids"]</code> (if using the latter example) instead of <code>SAMPLES</code>, as in:</p> <div class=highlight><pre><span></span><code>expand<span class=o>(</span><span class=s2>&quot;intermediate/{sample_id}_fastqc.zip&quot;</span>,
<span class=w>            </span><span class=nv>sample_id</span><span class=w> </span><span class=o>=</span><span class=w> </span>config<span class=o>[</span><span class=s2>&quot;sample_ids&quot;</span><span class=o>])</span>
</code></pre></div> <p>Do a dry-run afterwards to make sure that everything works as expected.</p> <p>The second point is trickier. Writing workflows in Snakemake is quite straightforward when the logic of the workflow is reflected in the file names, <em>i.e.</em> <code>my_sample.trimmed.deduplicated.sorted.fastq</code>, but that isn't always the case. In our case we have the FTP paths to the genome sequence and annotation where the naming doesn't quite fit with the rest of the workflow. The easiest solution is probably to make three parameters to hold these values, say <code>genome_id</code>, <code>genome_fasta_path</code> and <code>genome_gff_path</code>, but we will go for a somewhat more complex but very useful alternative. We want to construct a dictionary where something that will be a wildcard in the workflow is the key and the troublesome name is the value. An example might make this clearer (this is also in <code>config.yml</code> in the finished version of the workflow under <code>tutorials/git/</code>). This is a nested dictionary where "genomes" is a key with another dictionary as value, which in turn has genome ids as keys and so on. The idea is that we have a wildcard in the workflow that takes the id of a genome as value (either "NCTC8325" or "ST398" in this case). The fasta and gff3 paths can then be retrieved based on the value of the wildcard.</p> <div class=highlight><pre><span></span><code><span class=nt>genomes</span><span class=p>:</span>
<span class=w>  </span><span class=nt>NCTC8325</span><span class=p>:</span>
<span class=w>    </span><span class=nt>fasta</span><span class=p>:</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/fasta/bacteria_18_collection/staphylococcus_aureus_subsp_aureus_nctc_8325/dna//Staphylococcus_aureus_subsp_aureus_nctc_8325.ASM1342v1.dna_rm.toplevel.fa.gz</span>
<span class=w>    </span><span class=nt>gff3</span><span class=p>:</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/gff3/bacteria_18_collection/staphylococcus_aureus_subsp_aureus_nctc_8325//Staphylococcus_aureus_subsp_aureus_nctc_8325.ASM1342v1.37.gff3.gz</span>
<span class=w>  </span><span class=nt>ST398</span><span class=p>:</span>
<span class=w>    </span><span class=nt>fasta</span><span class=p>:</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/fasta/bacteria_18_collection//staphylococcus_aureus_subsp_aureus_st398/dna/Staphylococcus_aureus_subsp_aureus_st398.ASM958v1.dna.toplevel.fa.gz</span>
<span class=w>    </span><span class=nt>gff3</span><span class=p>:</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/gff3/bacteria_18_collection/staphylococcus_aureus_subsp_aureus_st398//Staphylococcus_aureus_subsp_aureus_st398.ASM958v1.37.gff3.gz</span>
</code></pre></div> <p>Go ahead and add the section above to <code>config.yml</code>.</p> <p>Let's now look at how to do the mapping from genome id to fasta path in the rule <code>get_genome_fasta</code>. This is how the rule currently looks (if you have added the log section as previously described).</p> <div class=highlight><pre><span></span><code><span class=n>rule</span> <span class=n>get_genome_fasta</span><span class=p>:</span>
<span class=w>    </span><span class=sd>&quot;&quot;&quot;</span>
<span class=sd>    Retrieve the sequence in fasta format for a genome.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>output</span><span class=p>:</span>
        <span class=s2>&quot;data/raw_external/NCTC8325.fa.gz&quot;</span>
    <span class=n>log</span><span class=p>:</span>
        <span class=s2>&quot;results/logs/get_genome_fasta/NCTC8325.log&quot;</span>
    <span class=n>shell</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;</span>
<span class=sd>        wget ftp://ftp.ensemblgenomes.org/pub/bacteria/release-37/fasta/bacteria_18_collection/staphylococcus_aureus_subsp_aureus_nctc_8325/dna//Staphylococcus_aureus_subsp_aureus_nctc_8325.ASM1342v1.dna_rm.toplevel.fa.gz -O {output} -o {log}</span>
<span class=sd>        &quot;&quot;&quot;</span>
</code></pre></div> <p>We don't want the hardcoded genome id <code>NCTC8325</code>, so replace that with a wildcard, say <code>{genome_id}</code> (remember to add the wildcard to the <code>log:</code> directive as well).</p> <p>Also change in <code>index_genome</code> to use a wildcard rather than a hardcoded genome id. Here you will run into a complication if you have followed the previous instructions and use the <code>expand()</code> expression. We want the list to expand to <code>["intermediate/{genome_id}.1.bt2", "intermediate/{genome_id}.2.bt2", ...]</code>, <em>i.e.</em> only expanding the wildcard referring to the bowtie2 index. To keep the <code>genome_id</code> wildcard from being expanded we have to "mask" it with double curly brackets: <code>`. In addition, we need to replace the hardcoded</code>intermediate/NCTC8325<code wildcards.genome_id=wildcards.genome_id>in the shell directive of the rule with the genome id wildcard. Inside the shell directive the wildcard object is accessed with this syntax:</code>`, so the bowtie2-build command should be:</p> <div class=highlight><pre><span></span><code>bowtie2-build<span class=w> </span>tempfile<span class=w> </span>intermediate/<span class=o>{</span>wildcards.genome_id<span class=o>}</span><span class=w> </span>&gt;<span class=w> </span><span class=o>{</span>log<span class=o>}</span>
</code></pre></div> <p>We now need to supply the remote paths to the fasta and gff files for a given genome id. Because we've added this information to the config file we just need to pass it to the rule in some way.</p> <p>Take a look at the code and <code>get_genome_fasta</code> rule below. Here we have defined a function called <code>get_fasta_path</code> which takes the <code>wildcards</code> object as its only argument. This object allows access to the wildcards values via attributes (here <code>wildcards.genome_id</code>). The function will then look in the nested <code>config</code> dictionary and return the value of the fasta path for the key <code>wildcards.genome_id</code>. In the rule this path is stored in the <code>fasta_path</code> param value and is made available to <code>wget</code> in the shell directive.</p> <div class=highlight><pre><span></span><code><span class=k>def</span> <span class=nf>get_fasta_path</span><span class=p>(</span><span class=n>wildcards</span><span class=p>):</span>
    <span class=k>return</span> <span class=n>config</span><span class=p>[</span><span class=s2>&quot;genomes&quot;</span><span class=p>][</span><span class=n>wildcards</span><span class=o>.</span><span class=n>genome_id</span><span class=p>][</span><span class=s2>&quot;fasta&quot;</span><span class=p>]</span>

<span class=n>rule</span> <span class=n>get_genome_fasta</span><span class=p>:</span>
<span class=w>    </span><span class=sd>&quot;&quot;&quot;</span>
<span class=sd>    Retrieve the sequence in fasta format for a genome.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>output</span><span class=p>:</span>
        <span class=s2>&quot;data/raw_external/</span><span class=si>{genome_id}</span><span class=s2>.fa.gz&quot;</span>
    <span class=n>log</span><span class=p>:</span>
        <span class=s2>&quot;results/logs/get_genome_fasta/</span><span class=si>{genome_id}</span><span class=s2>.log&quot;</span>
    <span class=n>params</span><span class=p>:</span>
        <span class=n>fasta_path</span> <span class=o>=</span> <span class=n>get_fasta_path</span>
    <span class=n>shell</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;</span>
<span class=sd>        wget {params.fasta_path} -O {output} -o {log}</span>
<span class=sd>        &quot;&quot;&quot;</span>
</code></pre></div> <p>Note that this will only work if the <code>{genome_id}</code> wildcard can be resolved to something defined in the config (currently <code>NCTC8325</code> or <code>ST398</code>). If you try to generate a fasta file for a genome id not defined in the config Snakemake will complain, even at the dry-run stage.</p> <p>Now change the <code>get_genome_gff3</code> rule in a similar manner.</p> <p>The rules <code>get_genome_fasta</code>, <code>get_genome_gff3</code> and <code>index_genome</code> can now download and index <em>any genome</em> as long as we provide valid links in the config file.</p> <p>However, we need to define somewhere which genome id we actually want to use when running the workflow. This needs to be done both in <code>align_to_genome</code> and <code>generate_count_table</code>. Do this by introducing a parameter in <code>config.yml</code> called <code>"genome_id"</code> (you can set it to either <code>NCTC8325</code> or <code>ST398</code>).</p> <p>Now we can resolve the <code>genome_id</code> wildcard from the config. See below for an example for <code>align_to_genome</code>. Here the <code>substr</code> wildcard gets expanded from a list while <code>genome_id</code> gets expanded from the config file.</p> <div class=highlight><pre><span></span><code><span class=nb>input</span><span class=p>:</span>
    <span class=n>index</span> <span class=o>=</span> <span class=n>expand</span><span class=p>(</span><span class=s2>&quot;intermediate/</span><span class=si>{genome_id}</span><span class=s2>.</span><span class=si>{substr}</span><span class=s2>.bt2&quot;</span><span class=p>,</span>
           <span class=n>genome_id</span> <span class=o>=</span> <span class=n>config</span><span class=p>[</span><span class=s2>&quot;genome_id&quot;</span><span class=p>],</span>
           <span class=n>substr</span> <span class=o>=</span> <span class=p>[</span><span class=s2>&quot;1&quot;</span><span class=p>,</span> <span class=s2>&quot;2&quot;</span><span class=p>,</span> <span class=s2>&quot;3&quot;</span><span class=p>,</span> <span class=s2>&quot;4&quot;</span><span class=p>,</span> <span class=s2>&quot;rev.1&quot;</span><span class=p>,</span> <span class=s2>&quot;rev.2&quot;</span><span class=p>])</span>
</code></pre></div> <p>Also change the hardcoded genome id in the <code>generate_count_table</code> input in a similar manner.</p> <p>In general, we want the rules as far downstream as possible in the workflow to be the ones that determine what the wildcards should resolve to. In our case this is <code>align_to_genome</code> and <code>generate_count_table</code>. You can think of it like the rule that really "needs" the file asks for it, and then it's up to Snakemake to determine how it can use all the available rules to generate it. Here the <code>align_to_genome</code> rule says "I need this genome index to align my sample to" and then it's up to Snakemake to determine how to download and build the index.</p> <p>One last thing is to change the hardcoded <code>NCTC8325</code> in the <code>shell:</code> directive of <code>align_to_genome</code>. Bowtie2 expects the index name supplied with the <code>-x</code> flag to be without the ".*.bt2" suffix so we can't use <code>-x {input.index}</code>. Instead we'll insert the genome_id directly from the config like this:</p> <div class=highlight><pre><span></span><code>shell:
<span class=w>    </span><span class=s2>&quot;&quot;&quot;  </span>
<span class=s2>    bowtie2 -x intermediate/{config[genome_id]} -U {input[0]} &gt; {output} 2&gt;{log}</span>
<span class=s2>    &quot;&quot;&quot;</span>
</code></pre></div> <div class="admonition success"> <p class=admonition-title>Quick recap</p> <p>In this section we've learned how to generalize a Snakemake workflow with a number of excellent features:</p> <ul> <li>A general RNA-seq pipeline which can easily be reused between projects, thanks to clear separation between code and settings.</li> <li>Great traceability due to logs and summary tables.</li> <li>Clearly defined the environment for the workflow using Conda.</li> <li>The workflow is neat and free from temporary files due to using <code>temp()</code> and <code>shadow</code>.</li> <li>A logical directory structure which makes it easy to separate raw data, intermediate files, and results.</li> <li>A project set up in a way that makes it very easy to distribute and reproduce either via Git, Snakemake's <code>--archive</code> option or a Docker image.</li> </ul> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../snakemake-9-shadow-rules/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Shadow rules" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Shadow rules </div> </div> </a> <a href=../snakemake-11-extra-material/ class="md-footer__link md-footer__link--next" aria-label="Next: Extra material" rel=next> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Extra material </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../../..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "header.autohide", "navigation.footer", "navigation.indexes", "navigation.tabs", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "toc.integrate"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../../../assets/javascripts/bundle.b4d07000.min.js></script> </body> </html>