<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Workshop on Tools for Reproducible Research"><meta name=author content="Jacques Dainat et al."><link href=https://southgreenplatform.github.io/training_reproducible_research/pages/jupyter/jupyter-8-the-mrsa-case-study/ rel=canonical><link href=../jupyter-7-converting-notebooks/ rel=prev><link href=../jupyter-9-extra-material/ rel=next><link rel=icon href=../../../assets/images/favicon.png><meta name=generator content="mkdocs-1.4.3, mkdocs-material-9.1.15"><title>The MRSA case study - Bioinformatics Workshop on Tools for Reproducible Research</title><link rel=stylesheet href=../../../assets/stylesheets/main.26e3688c.min.css><link rel=stylesheet href=../../../assets/stylesheets/palette.ecc896b0.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../../stylesheets/extra.css><script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#sharing-your-work class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-header__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../images/southgreen.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bioinformatics Workshop on Tools for Reproducible Research </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> The MRSA case study </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../../introduction/introduction_southgreen/ class="md-tabs__link md-tabs__link--active"> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../cheat_sheet/bash/bash/ class=md-tabs__link> Cheat sheet </a> </li> <li class=md-tabs__item> <a href=../../slido/ class=md-tabs__link> Ask a question </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-nav__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../images/southgreen.png alt=logo> </a> Bioinformatics Workshop on Tools for Reproducible Research </label> <div class=md-nav__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../.. class=md-nav__link> Home </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Tutorials <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../introduction/introduction_southgreen/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../introduction/roundtable/ class=md-nav__link> Roundtable Introduce Yourself </a> </li> <li class=md-nav__item> <a href=../../introduction/introduction_rr/ class=md-nav__link> Introduction Reproducible Research </a> </li> <li class=md-nav__item> <a href=../../data-management/data-management/ class=md-nav__link> Data Management </a> </li> <li class=md-nav__item> <a href=../../ice-breaker/ class=md-nav__link> Ice breaker session </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_6> <label class=md-nav__link for=__nav_2_6 id=__nav_2_6_label tabindex=0> Git <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_6_label aria-expanded=false> <label class=md-nav__title for=__nav_2_6> <span class="md-nav__icon md-icon"></span> Git </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../git/git-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../git/git-setup/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../git/git-2-creating-repositories/ class=md-nav__link> Creating a git repository </a> </li> <li class=md-nav__item> <a href=../../git/git-3-committing-changes/ class=md-nav__link> Committing changes </a> </li> <li class=md-nav__item> <a href=../../git/git-undoing-and-recovering/ class=md-nav__link> Undoing and recovering </a> </li> <li class=md-nav__item> <a href=../../git/git-4-ignoring-files/ class=md-nav__link> Ignoring files </a> </li> <li class=md-nav__item> <a href=../../git/git-5-branches/ class=md-nav__link> Branching and merging </a> </li> <li class=md-nav__item> <a href=../../git/git-6-tags/ class=md-nav__link> Tagging commits </a> </li> <li class=md-nav__item> <a href=../../git/git-7-working-remotely/ class=md-nav__link> Working remotly </a> </li> <li class=md-nav__item> <a href=../../git/git-8-conflicts/ class=md-nav__link> Conflict </a> </li> <li class=md-nav__item> <a href=../../git/git-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_7> <label class=md-nav__link for=__nav_2_7 id=__nav_2_7_label tabindex=0> Conda <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_7_label aria-expanded=false> <label class=md-nav__title for=__nav_2_7> <span class="md-nav__icon md-icon"></span> Conda </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../conda/conda-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../conda/conda-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../conda/conda-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../conda/conda-3-projects/ class=md-nav__link> Projects </a> </li> <li class=md-nav__item> <a href=../../conda/conda-4-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8> <label class=md-nav__link for=__nav_2_8 id=__nav_2_8_label tabindex=0> Workflow Manager System <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_8_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Workflow Manager System </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../wms/ class=md-nav__link> Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_2> <label class=md-nav__link for=__nav_2_8_2 id=__nav_2_8_2_label tabindex=0> Snakemake <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8_2> <span class="md-nav__icon md-icon"></span> Snakemake </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../snakemake/snakemake-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-3-visualising-workflows/ class=md-nav__link> Visualising workflow </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-4-the-mrsa-workflow/ class=md-nav__link> The MRSA Workflow </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-5-parameters/ class=md-nav__link> Parameters </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-6-logs/ class=md-nav__link> Logs </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-7-temporary-files/ class=md-nav__link> Temporary files </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-8-targets/ class=md-nav__link> Targets </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-9-shadow-rules/ class=md-nav__link> Shadow rules </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-10-generalizing-workflows/ class=md-nav__link> Generalizing workflows </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-11-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_3> <label class=md-nav__link for=__nav_2_8_3 id=__nav_2_8_3_label tabindex=0> Nextflow <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_3_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8_3> <span class="md-nav__icon md-icon"></span> Nextflow </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../nextflow/nextflow-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-nfcore/ class=md-nav__link> Getting started with nf-core </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-3-executing-workflows/ class=md-nav__link> Executing workflows </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-4-working-with-processes/ class=md-nav__link> Working with process </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-5-workflow-configuration/ class=md-nav__link> Workflow configuration </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-6-optimising-the-mrsa-workflow/ class=md-nav__link> Optmising the MRSA workflow </a> </li> <li class=md-nav__item> <a href=../../nextflow/nextflow-7-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_9> <label class=md-nav__link for=__nav_2_9 id=__nav_2_9_label tabindex=0> Containers <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_9_label aria-expanded=false> <label class=md-nav__title for=__nav_2_9> <span class="md-nav__icon md-icon"></span> Containers </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../containers/containers-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../containers/containers-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../containers/containers-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../containers/containers-3-building-images/ class=md-nav__link> Building a Docker image </a> </li> <li class=md-nav__item> <a href=../../containers/containers-4-managing-containers/ class=md-nav__link> Managing containers </a> </li> <li class=md-nav__item> <a href=../../containers/containers-5-sharing-images/ class=md-nav__link> Distributing your images </a> </li> <li class=md-nav__item> <a href=../../containers/containers-6-packaging-the-case-study/ class=md-nav__link> Packaging the case study </a> </li> <li class=md-nav__item> <a href=../../containers/containers-7-singularity/ class=md-nav__link> Singularity </a> </li> <li class=md-nav__item> <a href=../../containers/containers-8-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_10> <label class=md-nav__link for=__nav_2_10 id=__nav_2_10_label tabindex=0> Notebook <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_10_label aria-expanded=false> <label class=md-nav__title for=__nav_2_10> <span class="md-nav__icon md-icon"></span> Notebook </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../notebook/notebook-context/ class=md-nav__link> Context </a> </li> <li class=md-nav__item> <a href=../../notebook/notebook-markdown/ class=md-nav__link> Markdown </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_11> <label class=md-nav__link for=__nav_2_11 id=__nav_2_11_label tabindex=0> R Markdown <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_11_label aria-expanded=false> <label class=md-nav__title for=__nav_2_11> <span class="md-nav__icon md-icon"></span> R Markdown </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-3-text/ class=md-nav__link> Text </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-4-code-chunks/ class=md-nav__link> Code chunks </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-5-the-yaml-header/ class=md-nav__link> The YAML header </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-6-rendering/ class=md-nav__link> Rendering </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-7-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-8-extra-material/ class=md-nav__link> Extra materiel </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_12 checked> <label class=md-nav__link for=__nav_2_12 id=__nav_2_12_label tabindex=0> Jupyter <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_12_label aria-expanded=true> <label class=md-nav__title for=__nav_2_12> <span class="md-nav__icon md-icon"></span> Jupyter </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../jupyter-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../jupyter-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../jupyter-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../jupyter-3-magics/ class=md-nav__link> Magics </a> </li> <li class=md-nav__item> <a href=../jupyter-4-plotting/ class=md-nav__link> Plotting </a> </li> <li class=md-nav__item> <a href=../jupyter-5-widgets/ class=md-nav__link> Widgets </a> </li> <li class=md-nav__item> <a href=../jupyter-6-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../jupyter-7-converting-notebooks/ class=md-nav__link> Converting notebooks </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> The MRSA case study <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> The MRSA case study </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#sharing-your-work class=md-nav__link> Sharing your work </a> </li> <li class=md-nav__item> <a href=#shared-interactive-notebooks class=md-nav__link> Shared interactive notebooks </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../jupyter-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_13> <label class=md-nav__link for=__nav_2_13 id=__nav_2_13_label tabindex=0> Quarto <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_13_label aria-expanded=false> <label class=md-nav__title for=__nav_2_13> <span class="md-nav__icon md-icon"></span> Quarto </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../quarto/quarto-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-3-the-rendering/ class=md-nav__link> Rendering </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-4-the-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-5-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-6-comparison/ class=md-nav__link> Comparison </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-7-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_14> <label class=md-nav__link for=__nav_2_14 id=__nav_2_14_label tabindex=0> Making a project reproducible <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_14_label aria-expanded=false> <label class=md-nav__title for=__nav_2_14> <span class="md-nav__icon md-icon"></span> Making a project reproducible </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../project/project-1-review/ class=md-nav__link> Review </a> </li> <li class=md-nav__item> <a href=../../project/project-2-good_praticies/ class=md-nav__link> Good praticies </a> </li> <li class=md-nav__item> <a href=../../project/project-3-and_now/ class=md-nav__link> And now ? </a> </li> <li class=md-nav__item> <a href=../../project/project-4-demo/ class=md-nav__link> To end on a high note ! </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../take_down/take_down/ class=md-nav__link> Take down </a> </li> <li class=md-nav__item> <a href=../../final_word/ class=md-nav__link> Final word </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> Cheat sheet <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Cheat sheet </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../cheat_sheet/bash/bash/ class=md-nav__link> Bash </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/git/git/ class=md-nav__link> Git </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/git/git_official/ class=md-nav__link> Git official </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/conda/conda/ class=md-nav__link> Conda </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/notebook/notebook/ class=md-nav__link> Notebook </a> </li> <li class=md-nav__item> <a href=https://dockerlabs.collabnix.com/docker/cheatsheet/ class=md-nav__link> Docker </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../slido/ class=md-nav__link> Ask a question </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/edit/main/docs/pages/jupyter/jupyter-8-the-mrsa-case-study.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg> </a> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/raw/main/docs/pages/jupyter/jupyter-8-the-mrsa-case-study.md title="View source of this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.15 8.15 0 0 1-1.23-2Z"/></svg> </a> <h1>The MRSA case study</h1> <p>As you might remember from the <a href=../../introduction/introduction_rr/ >intro</a>, we are attempting to understand how lytic bacteriophages can be used as a future therapy for the multiresistant bacteria MRSA (methicillin-resistant <em>Staphylococcus aureus</em>). We have already seen how to define the project environment in the <a href=../../conda/conda-3-projects>Conda tutorial</a> and how to set up the workflow in the <a href=../../snakemake/snakemake-10-generalizing-workflows>Snakemake tutorial</a>. Here we explore the results from the Snakemake tutorial and generate a Supplementary Material file with some basic stats.</p> <p>In the <code>training-reproducible-research/tutorials/jupyter/</code> directory you will find a notebook called <code>supplementary_material.ipynb</code>. Open this notebook with Jupyter by running:</p> <div class=highlight><pre><span></span><code>jupyter<span class=w> </span>notebook<span class=w> </span>supplementary_material.ipynb
</code></pre></div> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>Using what you've learned about markdown in notebooks, add headers and descriptive text to subdivide sections as you add them. This will help you train how to structure and keep note of your work with a notebook.</p> </div> <p>You will see that the notebook contains only a little markdown text and a code cell with a function <code>get_geodata</code>. We'll start by adding a cell with some import statements.</p> <p>Run the cell with the <code>get_geodata</code> function and add a new cell directly after it. Then add the following to the new cell:</p> <div class=highlight><pre><span></span><code><span class=kn>import</span> <span class=nn>pandas</span> <span class=k>as</span> <span class=nn>pd</span>
<span class=kn>import</span> <span class=nn>seaborn</span> <span class=k>as</span> <span class=nn>sns</span>
<span class=kn>import</span> <span class=nn>matplotlib.pyplot</span> <span class=k>as</span> <span class=nn>plt</span>
<span class=kn>import</span> <span class=nn>numpy</span> <span class=k>as</span> <span class=nn>np</span>
</code></pre></div> <p>This imports the <code>pandas</code> (for working with tables), <code>seaborn</code> and <code>matplotlib.pyplot</code> (for plotting) and <code>numpy</code> (for numerical operations) Python modules.</p> <p>Also add:</p> <div class=highlight><pre><span></span><code><span class=kn>import</span> <span class=nn>matplotlib_inline</span>
<span class=n>matplotlib_inline</span><span class=o>.</span><span class=n>backend_inline</span><span class=o>.</span><span class=n>set_matplotlib_formats</span><span class=p>(</span><span class=s1>&#39;pdf&#39;</span><span class=p>,</span> <span class=s1>&#39;svg&#39;</span><span class=p>)</span>
</code></pre></div> <p>to set high-quality output for plots.</p> <p>Run the cell and create a new one below it.</p> <p>In the next cell we'll define some parameters to use for the notebook:</p> <div class=highlight><pre><span></span><code><span class=n>counts_file</span><span class=o>=</span><span class=s2>&quot;results/tables/counts.tsv&quot;</span>
<span class=n>summary_file</span><span class=o>=</span><span class=s2>&quot;results/tables/counts.tsv.summary&quot;</span>
<span class=n>multiqc_file</span><span class=o>=</span><span class=s2>&quot;intermediate/multiqc_general_stats.txt&quot;</span>
<span class=n>rulegraph_file</span><span class=o>=</span><span class=s2>&quot;results/rulegraph.png&quot;</span>
<span class=n>SRR_IDs</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;SRR935090&quot;</span><span class=p>,</span><span class=s2>&quot;SRR935091&quot;</span><span class=p>,</span><span class=s2>&quot;SRR935092&quot;</span><span class=p>]</span>
<span class=n>GSM_IDs</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;GSM1186459&quot;</span><span class=p>,</span><span class=s2>&quot;GSM1186460&quot;</span><span class=p>,</span><span class=s2>&quot;GSM1186461&quot;</span><span class=p>]</span>
<span class=n>GEO_ID</span><span class=o>=</span><span class=s2>&quot;GSE48896&quot;</span>
</code></pre></div> <p>As you can see we add paths to results files and define lists with some sample IDS. Run this cell and add a new one below it.</p> <p>Next, we'll fetch some sample information from NCBI using the <code>get_geodata</code> function defined at the start of the notebook and collate it into a dataframe.</p> <div class=highlight><pre><span></span><code><span class=n>id_df</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>DataFrame</span><span class=p>(</span><span class=n>data</span><span class=o>=</span><span class=n>GSM_IDs</span><span class=p>,</span> <span class=n>index</span><span class=o>=</span><span class=n>SRR_IDs</span><span class=p>,</span> <span class=n>columns</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;geo_accession&quot;</span><span class=p>])</span>
<span class=n>geo_df</span> <span class=o>=</span> <span class=n>get_geodata</span><span class=p>(</span><span class=n>GEO_ID</span><span class=p>)</span>
<span class=n>name_df</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>merge</span><span class=p>(</span><span class=n>id_df</span><span class=p>,</span> <span class=n>geo_df</span><span class=p>,</span> <span class=n>left_on</span><span class=o>=</span><span class=s2>&quot;geo_accession&quot;</span><span class=p>,</span> <span class=n>right_index</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
<span class=c1># Create a dictionary to rename sample ids in downstream plots</span>
<span class=n>name_dict</span> <span class=o>=</span> <span class=n>name_df</span><span class=o>.</span><span class=n>to_dict</span><span class=p>()</span>
</code></pre></div> <p>Run the cell and take a look at the contents of the <code>name_df</code> dataframe (<em>e.g.</em> run a cell with that variable only to output it below the cell).</p> <p>Now we'll load some statistics from the QC part of the workflow, specifically the 'general_stats' file from <code>multiqc</code>. Add the following to a new cell and run it:</p> <div class=highlight><pre><span></span><code><span class=n>qc</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=n>multiqc_file</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>)</span>
<span class=n>qc</span><span class=o>.</span><span class=n>rename</span><span class=p>(</span><span class=n>columns</span><span class=o>=</span><span class=k>lambda</span> <span class=n>x</span><span class=p>:</span> <span class=n>x</span><span class=o>.</span><span class=n>replace</span><span class=p>(</span><span class=s2>&quot;FastQC_mqc-generalstats-fastqc-&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>replace</span><span class=p>(</span><span class=s2>&quot;_&quot;</span><span class=p>,</span> <span class=s2>&quot; &quot;</span><span class=p>),</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
<span class=n>qc</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>merge</span><span class=p>(</span><span class=n>qc</span><span class=p>,</span> <span class=n>name_df</span><span class=p>,</span> <span class=n>left_on</span><span class=o>=</span><span class=s2>&quot;Sample&quot;</span><span class=p>,</span> <span class=n>right_index</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
<span class=n>qc</span>
</code></pre></div> <p>In the code above we load the multiqc file, rename the columns by stripping the <code>FastQC_mqc-generalstats-fastqc-</code> part from column names and replace underscores with spaces. Finally the table is merged with the information obtained in the step above and output to show summary statistics from the QC stage.</p> <p>Next it's time to start loading gene count results from the workflow. Start by reading the counts and summary results, then edit the columns and index:</p> <div class=highlight><pre><span></span><code><span class=c1># Read count data</span>
<span class=n>counts</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=n>counts_file</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=mi>0</span><span class=p>,</span> <span class=n>comment</span><span class=o>=</span><span class=s2>&quot;#&quot;</span><span class=p>,</span> <span class=n>index_col</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span>
<span class=c1># Read summary data</span>
<span class=n>counts_summary</span> <span class=o>=</span> <span class=n>pd</span><span class=o>.</span><span class=n>read_csv</span><span class=p>(</span><span class=n>summary_file</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>index_col</span><span class=o>=</span><span class=mi>0</span><span class=p>)</span>
<span class=c1># Rename columns to extract SRR ids</span>
<span class=n>counts</span><span class=o>.</span><span class=n>rename</span><span class=p>(</span><span class=n>columns</span> <span class=o>=</span> <span class=k>lambda</span> <span class=n>x</span><span class=p>:</span> <span class=n>x</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=s2>&quot;/&quot;</span><span class=p>)[</span><span class=o>-</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>replace</span><span class=p>(</span><span class=s2>&quot;.sorted.bam&quot;</span><span class=p>,</span><span class=s2>&quot;&quot;</span><span class=p>),</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
<span class=n>counts_summary</span><span class=o>.</span><span class=n>rename</span><span class=p>(</span><span class=n>columns</span> <span class=o>=</span> <span class=k>lambda</span> <span class=n>x</span><span class=p>:</span> <span class=n>x</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=s2>&quot;/&quot;</span><span class=p>)[</span><span class=o>-</span><span class=mi>1</span><span class=p>]</span><span class=o>.</span><span class=n>replace</span><span class=p>(</span><span class=s2>&quot;.sorted.bam&quot;</span><span class=p>,</span><span class=s2>&quot;&quot;</span><span class=p>),</span> <span class=n>inplace</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
</code></pre></div> <p>Take a look at the <code>counts</code> dataframe to get an idea of the data structure. As you can see the dataframe shows genes as rows while the columns shows various information such as start and stop, strand and length of the genes. The last three columns contain counts of the genes in each of the samples.</p> <p>If you have a look at the <code>counts_summary</code> dataframe you will see statistics from the read assignment step, showing number of reads that could be properly assigned as well as number of reads that could not be assigned to genes for various reasons.</p> <p>Now let's generate a barplot of the summary statistics. Before we plot, we'll remove rows that have only zero values:</p> <div class=highlight><pre><span></span><code><span class=c1># Remove rows with only zero values</span>
<span class=n>summary_plot_data</span> <span class=o>=</span> <span class=n>counts_summary</span><span class=o>.</span><span class=n>loc</span><span class=p>[</span><span class=n>counts_summary</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>1</span><span class=p>)</span><span class=o>&gt;</span><span class=mi>0</span><span class=p>]</span>
</code></pre></div> <p>Now for the plotting:</p> <div class=highlight><pre><span></span><code><span class=c1># Set color palette to &#39;Set2&#39;</span>
<span class=n>colors</span> <span class=o>=</span> <span class=n>sns</span><span class=o>.</span><span class=n>color_palette</span><span class=p>(</span><span class=s2>&quot;Set2&quot;</span><span class=p>)</span>
<span class=c1># Create a stacked barplot</span>
<span class=n>ax</span> <span class=o>=</span> <span class=n>summary_plot_data</span><span class=o>.</span><span class=n>T</span><span class=o>.</span><span class=n>plot</span><span class=p>(</span><span class=n>kind</span><span class=o>=</span><span class=s2>&quot;bar&quot;</span><span class=p>,</span> <span class=n>stacked</span><span class=o>=</span><span class=kc>True</span><span class=p>,</span> <span class=n>color</span><span class=o>=</span><span class=n>colors</span><span class=p>)</span>
<span class=c1># Move legend and set legend title</span>
<span class=n>ax</span><span class=o>.</span><span class=n>legend</span><span class=p>(</span><span class=n>bbox_to_anchor</span><span class=o>=</span><span class=p>(</span><span class=mi>1</span><span class=p>,</span><span class=mi>1</span><span class=p>),</span> <span class=n>title</span><span class=o>=</span><span class=s2>&quot;Category&quot;</span><span class=p>);</span>
</code></pre></div> <p>The final plot will be a heatmap of gene counts for a subset of the genes. We'll select genes whose standard deviation/mean count across samples is greater than 1.5, <strong>and</strong> have a maximum of at least 5 reads in 1 or more sample:</p> <div class=highlight><pre><span></span><code><span class=c1># Slice the dataframe to only sample counts</span>
<span class=n>count_data</span> <span class=o>=</span> <span class=n>counts</span><span class=o>.</span><span class=n>loc</span><span class=p>[:,</span> <span class=n>SRR_IDs</span><span class=p>]</span>
<span class=c1># Filter to genes with std/mean &gt; 1.2 and with a max of at least 5</span>
<span class=n>heatmap_data</span> <span class=o>=</span> <span class=n>count_data</span><span class=o>.</span><span class=n>loc</span><span class=p>[(</span><span class=n>count_data</span><span class=o>.</span><span class=n>std</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>div</span><span class=p>(</span><span class=n>count_data</span><span class=o>.</span><span class=n>mean</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>1</span><span class=p>))</span><span class=o>&gt;</span><span class=mf>1.2</span><span class=p>)</span><span class=o>&amp;</span><span class=p>(</span><span class=n>count_data</span><span class=o>.</span><span class=n>max</span><span class=p>(</span><span class=n>axis</span><span class=o>=</span><span class=mi>1</span><span class=p>)</span><span class=o>&gt;</span><span class=mi>5</span><span class=p>)]</span>
</code></pre></div> <p>We'll also replace the SRR ids with the title of samples used in the study, using the <code>name_dict</code> dictionary created further up in the notebook:</p> <div class=highlight><pre><span></span><code><span class=n>heatmap_data</span> <span class=o>=</span> <span class=n>heatmap_data</span><span class=o>.</span><span class=n>rename</span><span class=p>(</span><span class=n>columns</span> <span class=o>=</span> <span class=n>name_dict</span><span class=p>[</span><span class=s1>&#39;title&#39;</span><span class=p>])</span>
</code></pre></div> <p>Now let's plot the heatmap. We'll log-transform the counts, set color scale to Blue-Yellow-Red and cluster both samples and genes using 'complete' linkage clustering:</p> <div class=highlight><pre><span></span><code><span class=k>with</span> <span class=n>sns</span><span class=o>.</span><span class=n>plotting_context</span><span class=p>(</span><span class=s2>&quot;notebook&quot;</span><span class=p>,</span> <span class=n>font_scale</span><span class=o>=</span><span class=mf>0.7</span><span class=p>):</span>
    <span class=n>ax</span> <span class=o>=</span> <span class=n>sns</span><span class=o>.</span><span class=n>clustermap</span><span class=p>(</span><span class=n>data</span><span class=o>=</span><span class=n>np</span><span class=o>.</span><span class=n>log10</span><span class=p>(</span><span class=n>heatmap_data</span><span class=o>+</span><span class=mi>1</span><span class=p>),</span> <span class=n>cmap</span><span class=o>=</span><span class=s2>&quot;RdYlBu_r&quot;</span><span class=p>,</span>
                        <span class=n>method</span><span class=o>=</span><span class=s2>&quot;complete&quot;</span><span class=p>,</span> <span class=n>yticklabels</span><span class=o>=</span><span class=kc>True</span><span class=p>,</span> <span class=n>linewidth</span><span class=o>=</span><span class=mf>.5</span><span class=p>,</span>
                        <span class=n>cbar_pos</span><span class=o>=</span><span class=p>(</span><span class=mf>.7</span><span class=p>,</span> <span class=mf>.85</span><span class=p>,</span> <span class=mf>.05</span><span class=p>,</span> <span class=mf>.1</span><span class=p>),</span> <span class=n>figsize</span><span class=o>=</span><span class=p>(</span><span class=mi>3</span><span class=p>,</span><span class=mi>9</span><span class=p>))</span>
    <span class=n>plt</span><span class=o>.</span><span class=n>setp</span><span class=p>(</span><span class=n>ax</span><span class=o>.</span><span class=n>ax_heatmap</span><span class=o>.</span><span class=n>get_xticklabels</span><span class=p>(),</span> <span class=n>rotation</span><span class=o>=</span><span class=mi>270</span><span class=p>)</span>
</code></pre></div> <p>In the code above we use the seaborn <code>plotting_context</code> function to scale all text elements of the heatmap in one go.</p> <p>As a final step we'll add some info for reproducibility under the <strong>Reproducibility</strong> section. To add the overview image of the workflow found in <code>results/rulegraph.png</code> we can use the <code>Image</code> function from <code>IPython.display</code>:</p> <div class=highlight><pre><span></span><code><span class=kn>from</span> <span class=nn>IPython.display</span> <span class=kn>import</span> <span class=n>Image</span>
<span class=n>Image</span><span class=p>(</span><span class=n>rulegraph_file</span><span class=p>)</span>
</code></pre></div> <p>Let's also output the full conda environment so that all packages and versions are included in the notebook. There are several ways this can be done, for example you could simply add:</p> <div class=highlight><pre><span></span><code><span class=err>!</span><span class=n>conda</span> <span class=nb>list</span>
</code></pre></div> <p>to the end of the notebook.</p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>If you want to know more about how notebooks can be integrated into Snakemake worfklows, see the Extra material at the end of this tutorial</p> </div> <h2 id=sharing-your-work>Sharing your work<a class=headerlink href=#sharing-your-work title="Permanent link">#</a></h2> <p>The files you're working with come from a GitHub repo. Both GitHub and Bitbucket can render Jupyter notebooks as well as other types of Markdown documents. Now go to our GitHub repo at <a href=https://github.com/SouthGreenPlatform/training_reproducible_research>https://github.com/SouthGreenPlatform/training_reproducible_research</a> and navigate to <code>tutorials/jupyter/supplementary_material.ipynb</code>.</p> <div align=center> <p><img alt src=../images/jupyter_supplementary.png width=700px></p> </div> <p>As you can imagine, having this very effortless way of sharing results can greatly increase the visibility of your work. You work as normal on your project, and push regularly to the repository as you would anyways, and the output is automatically available for anyone to see. Or for a select few if you're not ready to share your findings with the world quite yet.</p> <p>Say your notebook isn't on Github/Bitbucket. All hope isn't lost there. Jupyter.org provides a neat functionality called <em>nbviewer</em>, where you can paste a URL to any notebook and they will render it for you. Go to <a href=https://nbviewer.jupyter.org>https://nbviewer.jupyter.org</a> and try this out with our notebook.</p> <div class=highlight><pre><span></span><code>https://raw.githubusercontent.com/NBISweden/workshop-reproducible-research/main/tutorials/jupyter/supplementary_material.ipynb
</code></pre></div> <h2 id=shared-interactive-notebooks>Shared interactive notebooks<a class=headerlink href=#shared-interactive-notebooks title="Permanent link">#</a></h2> <p>So far we've only shared static representations of notebooks. A strong trend at the moment is to run your notebooks in the cloud, so that the person you want to share with could actually execute and modify your code. This is a great way of increasing visibility and letting collaborators or readers get more hands-on with your data and analyses. From a reproducibility perspective, there are both advantages and drawbacks. On the plus side is that running your work remotely forces you to be strict when it comes to defining the environment it uses (probably in the form of a Conda environment or Docker image). On the negative side is that you become reliant on a third-party service that might change input formats, go out of business, or change payment model.</p> <p>Here we will try out a service called Binder, which lets you run and share Jupyter Notebooks in Git repositories for free. There are a number of <a href=https://github.com/binder-examples/ >example repositories</a> that are setup to be used with Binder. Navigate to <a href=https://github.com/binder-examples/conda/ >https://github.com/binder-examples/conda/</a> to see one such example. As you can see the repository contains a LICENSE file, a README, an environment file and a notebook. To use a repository with Binder the environment file should contain all the packages needed to run notebooks in the repo. So let's try to run the <code>index.ipynb</code> file using Binder:</p> <p>Just go to <a href=https://mybinder.org>https://mybinder.org</a> and paste the link to the GitHub repo. Note the link that you can use to share your notebook. Then press "launch".</p> <div align=center> <p><img alt src=../images/binder.png width=700px></p> </div> <p>What will happen now is that:</p> <ul> <li>Binder detects the <code>environment.yml</code> file in the root of the repo. Binder then builds a <em>Docker image</em> based on the file. This might take a minute or two. You can follow the progress in the build log.</li> <li>Binder then launches the Jupyter Notebook server in the Docker container..</li> <li>..and opens a browser tab with it for you.</li> </ul> <p>Once the process is finished you will be presented with a Jupyter server overview of the contents in the repository. Click on the <code>index.ipynb</code> notebook to open it. Tada! You are now able to interact with (and modify) someone else's notebook online.</p> <p>Applied to your own projects you now have a way to run analyses in the cloud and in an environment that you define yourself. All that's needed for someone to replicate your analyses is that you share a link with them. Note that notebooks on Binder are read-only; its purpose is for trying out and showing existing notebooks rather than making new ones.</p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>By default Binder looks for configuration files such as environment.yml in the root of the repository being built. But you may also put such files outside the root by making a <code>binder/</code> folder in the root and placing the file there. </p> </div> <div class="admonition info"> <p class=admonition-title>A note on transparency</p> <p>Resources like Github/Bitbucket and Jupyter Notebooks have changed the way we do scientific research by encouraging visibility, social interaction and transparency. It was not long ago that the analysis scripts and workflows in a lab were well-guarded secrets that we only most reluctantly shared with others. Assuming that it was even possible. In most cases, the only postdoc who knew how to get it to work had left for a new position in industry, or no one could remember the password to the file server. If you're a PhD student, we encourage you to embrace this new development wholeheartedly, for it will make your research better and make you into a better scientist. And you will have more fun.</p> </div> <div class="admonition success"> <p class=admonition-title>Quick recap</p> <p>In this section we've learned:</p> <ul> <li>How notebooks can be used to generate summary statistics and plots using the results of a workflow run</li> <li>How to share notebooks via <a href=https://nbviewer.jupyter.org>nbviewer</a> and <a href=https://mybinder.org>Binder</a></li> </ul> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../jupyter-7-converting-notebooks/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Converting notebooks" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Converting notebooks </div> </div> </a> <a href=../jupyter-9-extra-material/ class="md-footer__link md-footer__link--next" aria-label="Next: Extra material" rel=next> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Extra material </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../../..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "header.autohide", "navigation.footer", "navigation.indexes", "navigation.tabs", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "toc.integrate"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../../../assets/javascripts/bundle.b4d07000.min.js></script> </body> </html>