<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Workshop on Tools for Reproducible Research"><meta name=author content="Jacques Dainat et al."><link href=https://southgreenplatform.github.io/training_reproducible_research/pages/nextflow/nextflow-7-extra-material/ rel=canonical><link href=../nextflow-6-optimising-the-mrsa-workflow/ rel=prev><link href=../../containers/containers-1-introduction/ rel=next><link rel=icon href=../../../assets/images/favicon.png><meta name=generator content="mkdocs-1.4.3, mkdocs-material-9.1.15"><title>Extra material - Bioinformatics Workshop on Tools for Reproducible Research</title><link rel=stylesheet href=../../../assets/stylesheets/main.26e3688c.min.css><link rel=stylesheet href=../../../assets/stylesheets/palette.ecc896b0.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../../stylesheets/extra.css><script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#using-containers-in-nextflow class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-header__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../images/southgreen.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bioinformatics Workshop on Tools for Reproducible Research </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Extra material </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../../introduction/introduction_southgreen/ class="md-tabs__link md-tabs__link--active"> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../cheat_sheet/bash/bash/ class=md-tabs__link> Cheat sheet </a> </li> <li class=md-tabs__item> <a href=../../slido/ class=md-tabs__link> Ask a question </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-nav__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../images/southgreen.png alt=logo> </a> Bioinformatics Workshop on Tools for Reproducible Research </label> <div class=md-nav__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../.. class=md-nav__link> Home </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Tutorials <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=true> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../introduction/introduction_southgreen/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../introduction/roundtable/ class=md-nav__link> Roundtable Introduce Yourself </a> </li> <li class=md-nav__item> <a href=../../introduction/introduction_rr/ class=md-nav__link> Introduction Reproducible Research </a> </li> <li class=md-nav__item> <a href=../../data-management/data-management/ class=md-nav__link> Data Management </a> </li> <li class=md-nav__item> <a href=../../ice-breaker/ class=md-nav__link> Ice breaker session </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_6> <label class=md-nav__link for=__nav_2_6 id=__nav_2_6_label tabindex=0> Git <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_6_label aria-expanded=false> <label class=md-nav__title for=__nav_2_6> <span class="md-nav__icon md-icon"></span> Git </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../git/git-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../git/git-setup/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../git/git-2-creating-repositories/ class=md-nav__link> Creating a git repository </a> </li> <li class=md-nav__item> <a href=../../git/git-3-committing-changes/ class=md-nav__link> Committing changes </a> </li> <li class=md-nav__item> <a href=../../git/git-undoing-and-recovering/ class=md-nav__link> Undoing and recovering </a> </li> <li class=md-nav__item> <a href=../../git/git-4-ignoring-files/ class=md-nav__link> Ignoring files </a> </li> <li class=md-nav__item> <a href=../../git/git-5-branches/ class=md-nav__link> Branching and merging </a> </li> <li class=md-nav__item> <a href=../../git/git-6-tags/ class=md-nav__link> Tagging commits </a> </li> <li class=md-nav__item> <a href=../../git/git-7-working-remotely/ class=md-nav__link> Working remotly </a> </li> <li class=md-nav__item> <a href=../../git/git-8-conflicts/ class=md-nav__link> Conflict </a> </li> <li class=md-nav__item> <a href=../../git/git-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_7> <label class=md-nav__link for=__nav_2_7 id=__nav_2_7_label tabindex=0> Conda <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_7_label aria-expanded=false> <label class=md-nav__title for=__nav_2_7> <span class="md-nav__icon md-icon"></span> Conda </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../conda/conda-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../conda/conda-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../conda/conda-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../conda/conda-3-projects/ class=md-nav__link> Projects </a> </li> <li class=md-nav__item> <a href=../../conda/conda-4-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8 checked> <label class=md-nav__link for=__nav_2_8 id=__nav_2_8_label tabindex=0> Workflow Manager System <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_8_label aria-expanded=true> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Workflow Manager System </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../wms/ class=md-nav__link> Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_2> <label class=md-nav__link for=__nav_2_8_2 id=__nav_2_8_2_label tabindex=0> Snakemake <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8_2> <span class="md-nav__icon md-icon"></span> Snakemake </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../snakemake/snakemake-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-3-visualising-workflows/ class=md-nav__link> Visualising workflow </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-4-the-mrsa-workflow/ class=md-nav__link> The MRSA Workflow </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-5-parameters/ class=md-nav__link> Parameters </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-6-logs/ class=md-nav__link> Logs </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-7-temporary-files/ class=md-nav__link> Temporary files </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-8-targets/ class=md-nav__link> Targets </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-9-shadow-rules/ class=md-nav__link> Shadow rules </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-10-generalizing-workflows/ class=md-nav__link> Generalizing workflows </a> </li> <li class=md-nav__item> <a href=../../snakemake/snakemake-11-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_3 checked> <label class=md-nav__link for=__nav_2_8_3 id=__nav_2_8_3_label tabindex=0> Nextflow <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_3_label aria-expanded=true> <label class=md-nav__title for=__nav_2_8_3> <span class="md-nav__icon md-icon"></span> Nextflow </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../nextflow-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../nextflow-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../nextflow-nfcore/ class=md-nav__link> Getting started with nf-core </a> </li> <li class=md-nav__item> <a href=../nextflow-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../nextflow-3-executing-workflows/ class=md-nav__link> Executing workflows </a> </li> <li class=md-nav__item> <a href=../nextflow-4-working-with-processes/ class=md-nav__link> Working with process </a> </li> <li class=md-nav__item> <a href=../nextflow-5-workflow-configuration/ class=md-nav__link> Workflow configuration </a> </li> <li class=md-nav__item> <a href=../nextflow-6-optimising-the-mrsa-workflow/ class=md-nav__link> Optmising the MRSA workflow </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <a href=./ class="md-nav__link md-nav__link--active"> Extra material </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_9> <label class=md-nav__link for=__nav_2_9 id=__nav_2_9_label tabindex=0> Containers <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_9_label aria-expanded=false> <label class=md-nav__title for=__nav_2_9> <span class="md-nav__icon md-icon"></span> Containers </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../containers/containers-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../containers/containers-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../containers/containers-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../containers/containers-3-building-images/ class=md-nav__link> Building a Docker image </a> </li> <li class=md-nav__item> <a href=../../containers/containers-4-managing-containers/ class=md-nav__link> Managing containers </a> </li> <li class=md-nav__item> <a href=../../containers/containers-5-sharing-images/ class=md-nav__link> Distributing your images </a> </li> <li class=md-nav__item> <a href=../../containers/containers-6-packaging-the-case-study/ class=md-nav__link> Packaging the case study </a> </li> <li class=md-nav__item> <a href=../../containers/containers-7-singularity/ class=md-nav__link> Singularity </a> </li> <li class=md-nav__item> <a href=../../containers/containers-8-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_10> <label class=md-nav__link for=__nav_2_10 id=__nav_2_10_label tabindex=0> Notebook <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_10_label aria-expanded=false> <label class=md-nav__title for=__nav_2_10> <span class="md-nav__icon md-icon"></span> Notebook </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../notebook/notebook-context/ class=md-nav__link> Context </a> </li> <li class=md-nav__item> <a href=../../notebook/notebook-markdown/ class=md-nav__link> Markdown </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_11> <label class=md-nav__link for=__nav_2_11 id=__nav_2_11_label tabindex=0> R Markdown <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_11_label aria-expanded=false> <label class=md-nav__title for=__nav_2_11> <span class="md-nav__icon md-icon"></span> R Markdown </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-3-text/ class=md-nav__link> Text </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-4-code-chunks/ class=md-nav__link> Code chunks </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-5-the-yaml-header/ class=md-nav__link> The YAML header </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-6-rendering/ class=md-nav__link> Rendering </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-7-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../rmarkdown/r-markdown-8-extra-material/ class=md-nav__link> Extra materiel </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_12> <label class=md-nav__link for=__nav_2_12 id=__nav_2_12_label tabindex=0> Jupyter <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_12_label aria-expanded=false> <label class=md-nav__title for=__nav_2_12> <span class="md-nav__icon md-icon"></span> Jupyter </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../jupyter/jupyter-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-3-magics/ class=md-nav__link> Magics </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-4-plotting/ class=md-nav__link> Plotting </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-5-widgets/ class=md-nav__link> Widgets </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-6-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-7-converting-notebooks/ class=md-nav__link> Converting notebooks </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-8-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../jupyter/jupyter-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_13> <label class=md-nav__link for=__nav_2_13 id=__nav_2_13_label tabindex=0> Quarto <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_13_label aria-expanded=false> <label class=md-nav__title for=__nav_2_13> <span class="md-nav__icon md-icon"></span> Quarto </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../quarto/quarto-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-3-the-rendering/ class=md-nav__link> Rendering </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-4-the-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-5-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-6-comparison/ class=md-nav__link> Comparison </a> </li> <li class=md-nav__item> <a href=../../quarto/quarto-7-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_14> <label class=md-nav__link for=__nav_2_14 id=__nav_2_14_label tabindex=0> Making a project reproducible <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_14_label aria-expanded=false> <label class=md-nav__title for=__nav_2_14> <span class="md-nav__icon md-icon"></span> Making a project reproducible </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../project/project-1-review/ class=md-nav__link> Review </a> </li> <li class=md-nav__item> <a href=../../project/project-2-good_praticies/ class=md-nav__link> Good praticies </a> </li> <li class=md-nav__item> <a href=../../project/project-3-and_now/ class=md-nav__link> And now ? </a> </li> <li class=md-nav__item> <a href=../../project/project-4-demo/ class=md-nav__link> To end on a high note ! </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../take_down/take_down/ class=md-nav__link> Take down </a> </li> <li class=md-nav__item> <a href=../../final_word/ class=md-nav__link> Final word </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> Cheat sheet <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Cheat sheet </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../cheat_sheet/bash/bash/ class=md-nav__link> Bash </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/git/git/ class=md-nav__link> Git </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/git/git_official/ class=md-nav__link> Git official </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/conda/conda/ class=md-nav__link> Conda </a> </li> <li class=md-nav__item> <a href=../../cheat_sheet/notebook/notebook/ class=md-nav__link> Notebook </a> </li> <li class=md-nav__item> <a href=https://dockerlabs.collabnix.com/docker/cheatsheet/ class=md-nav__link> Docker </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../slido/ class=md-nav__link> Ask a question </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/edit/main/docs/pages/nextflow/nextflow-7-extra-material.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg> </a> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/raw/main/docs/pages/nextflow/nextflow-7-extra-material.md title="View source of this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.15 8.15 0 0 1-1.23-2Z"/></svg> </a> <p>There are many more things you can do with Nextflow than covered here. If you are interested to learn more details about Nextflow, we will briefly show some of its advanced features in this section. But first, here are some links to additional resources on Nextflow:</p> <ul> <li><a href=http://nextflow-io.github.io/patterns/index.html>Nextflow patterns</a> that can help with common operations and concepts</li> <li>The Nextflow <a href=https://www.nextflow.io/docs/latest/index.html>documentation</a></li> <li><a href=https://www.nextflow.io/blog/2020/learning-nextflow-in-2020.html>Learning Nextflow in 2020</a></li> <li>Nextflow training at <a href=https://seqera.io/training/ >Seqera</a></li> <li>A work-in-progress <a href=https://carpentries-incubator.github.io/workflows-nextflow/index.html>Nextflow Carpentry course</a></li> <li>Community help from <a href=https://join.slack.com/t/nextflow/shared_invite/zt-11iwlxtw5-R6SNBpVksOJAx5sPOXNrZg>Nextflow's Slack channel</a></li> </ul> <h1 id=using-containers-in-nextflow>Using containers in Nextflow<a class=headerlink href=#using-containers-in-nextflow title="Permanent link">#</a></h1> <p>Nextflow has built-in support for using both Docker and Singularity containers (and others too), either with a single container for the workflow as a whole or separate containers for each individual process. The simplest way to do it is to have a single container for your entire workflow, in which case you simply run the workflow and specify the image you want to use, like so:</p> <div class=highlight><pre><span></span><code><span class=c1># Run with docker</span>
nextflow<span class=w> </span>run<span class=w> </span>main.nf<span class=w> </span>-with-docker<span class=w> </span><span class=o>[</span>image<span class=o>]</span>

<span class=c1># Run with Singularity</span>
nextflow<span class=w> </span>run<span class=w> </span>main.nf<span class=w> </span>-with-singularity<span class=w> </span><span class=o>[</span>image<span class=o>]</span>.sif
</code></pre></div> <p>If you don't want to supply this at every execution, you can also add it directly to your configuration file:</p> <div class=highlight><pre><span></span><code><span class=c1># Docker configuration</span>
process.container<span class=w> </span><span class=o>=</span><span class=w> </span><span class=s1>&#39;image&#39;</span>
docker.enabled<span class=w> </span><span class=o>=</span><span class=w> </span><span class=nb>true</span>

<span class=c1># Singularity configuration</span>
process.container<span class=w> </span><span class=o>=</span><span class=w> </span><span class=s1>&#39;path/to/image.sif&#39;</span>
singularity.enabled<span class=w> </span><span class=o>=</span><span class=w> </span><span class=nb>true</span>
</code></pre></div> <p>If you instead would like to have each process use a different container you can use the <code>container</code> directive in your processes:</p> <div class=highlight><pre><span></span><code><span class=n>process</span><span class=w> </span><span class=n>PROCESS_01</span><span class=w> </span><span class=o>{</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=w>    </span><span class=nl>container:</span><span class=w> </span><span class=s1>&#39;image_01&#39;</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=o>}</span>

<span class=n>process</span><span class=w> </span><span class=n>PROCESS_02</span><span class=w> </span><span class=o>{</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=w>    </span><span class=nl>container:</span><span class=w> </span><span class=s1>&#39;image_02&#39;</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=o>}</span>
</code></pre></div> <p>Regardless of which solution you go for, Nextflow will execute all the processes inside the specified container. In practice, this means that Nextflow will automatically wrap your processes and run them by executing the Docker or Singularity command with the image you have provided.</p> <h1 id=using-conda-in-nextflow>Using Conda in Nextflow<a class=headerlink href=#using-conda-in-nextflow title="Permanent link">#</a></h1> <p>While you can execute Nextflow inside Conda environments just like you would any other type of software, you can also use Conda with Nextflow in the same way as for Docker and Singularity above. You can either supply an <code>environment.yml</code> file, the path to an existing environment or the packages and their versions directly in the <code>conda</code> directive, like so:</p> <div class=highlight><pre><span></span><code><span class=n>process</span><span class=w> </span><span class=n>PROCESS_01</span><span class=w> </span><span class=o>{</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=w>    </span><span class=nl>conda:</span><span class=w> </span><span class=s1>&#39;mrsa-environment.yml&#39;</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=o>}</span>
<span class=n>process</span><span class=w> </span><span class=n>PROCESS_02</span><span class=w> </span><span class=o>{</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=w>    </span><span class=nl>conda:</span><span class=w> </span><span class=s1>&#39;path/to/mrsa-env&#39;</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=o>}</span>
<span class=n>process</span><span class=w> </span><span class=n>PROCESS_03</span><span class=w> </span><span class=o>{</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=w>    </span><span class=nl>conda:</span><span class=w> </span><span class=s1>&#39;bioconda::bwa=0.7.17 bioconda::samtools=1.13&#39;</span>
<span class=w>    </span><span class=o>(...)</span>
<span class=o>}</span>
</code></pre></div> <p>You can use either of the methods described above with your configuration file as well, here exemplified using an <code>environment.yml</code> file:</p> <div class=highlight><pre><span></span><code><span class=n>process</span><span class=o>.</span><span class=na>conda</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s1>&#39;mrsa-environment.yml&#39;</span>
</code></pre></div> <h1 id=running-nextflow-on-hpc>Running Nextflow on HPC<a class=headerlink href=#running-nextflow-on-hpc title="Permanent link">#</a></h1> <p>A lot of researchers use HPC cluster, which is easily handled by Nextflow. What you need to do is to add an appropriate <em>profile</em> to your <code>nextflow.config</code> file. Example for iTrop:</p> <div class=highlight><pre><span></span><code>profiles {

    // Itrop general profile
    itrop {
        process {
            executor       = &#39;slurm&#39;
            clusterOptions = &quot;-A &#39;${params.account}&#39;&quot;
            memory         = { 6.GB * task.attempt }
            cpus           = { 1 * task.attempt }
            time           = { 10.h * task.attempt }
            scratch        = &#39;/scracth/$USER&#39;
            errorStrategy  = &#39;retry&#39;
            maxRetries     = 1
        }
    }
}
</code></pre></div> <p>This will add a profile to your workflow, which you can access by running the workflow with <code>-profile itrop</code>. You can leave the profile as-is, unless you want to tinker with <em>e.g.</em> compute resource specifications. That's all you need! Nextflow will take care of communications with SLURM (the system used by iTrop, specified by the <code>executor</code> line) and will send off jobs to the cluster for you, and everything will look exactly the same way as if you were executing the pipeline locally.</p> <p>The <code>memory</code>, <code>cpus</code> and <code>time</code> lines define the various resources Nextflow will use as well as how much to automatically increase them by if re-trying failed tasks; this, in turn, is specified by the <code>errorStrategy</code> and <code>maxRetries</code> variables. The <code>scratch</code> defines where each node's local storage is situated, which gives Nextflow the most optimal access to the cluster file system for temporary files.</p> <h1 id=advanced-channel-creation>Advanced channel creation<a class=headerlink href=#advanced-channel-creation title="Permanent link">#</a></h1> <p>The input data shown in the MRSA example workflow is not that complex, but Nextflow channels can do much more than that. A common scenario in high-throughput sequencing is that you have pairs of reads for each sample. Nextflow has a special, built-in way to create channels for this data type: the <code>fromFilePairs</code> channel factory:</p> <div class=highlight><pre><span></span><code><span class=n>Channel</span>
<span class=w>    </span><span class=o>.</span><span class=na>fromFilePairs</span><span class=w> </span><span class=o>(</span><span class=w> </span><span class=s2>&quot;data/*_R{1,2}.fastq.gz&quot;</span><span class=w> </span><span class=o>)</span>
<span class=w>    </span><span class=o>.</span><span class=na>set</span><span class=w>           </span><span class=o>{</span><span class=w> </span><span class=n>ch_raw_reads</span><span class=w> </span><span class=o>}</span>
</code></pre></div> <p>This will create a channel containing all the reads in the <code>data/</code> directory in the format <code>&lt;sample&gt;_R1.fastq.gz</code> and <code>&lt;sample&gt;_R2.fastq.gz</code> and will pair them together into a nested tuple looking like this:</p> <div class=highlight><pre><span></span><code><span class=o>[</span><span class=n>sample</span><span class=o>,</span><span class=w> </span><span class=o>[</span><span class=n>data</span><span class=s>/sample_R1.fastq.gz, data/</span><span class=n>sample_R2</span><span class=o>.</span><span class=na>fastq</span><span class=o>.</span><span class=na>gz</span><span class=o>]]</span>
</code></pre></div> <p>The first element of the tuple (<code>[0]</code>) thus contains the value <code>sample</code>, while the second element (<code>[1]</code>) contains another tuple with paths to both read files. This nested tuple can be passed into processes for <em>e.g.</em> read alignment, and it makes the entire procedure of going from read pairs (<em>i.e.</em> two separate files, one sample) into a single alignment file (one file, one sample) very simple. For more methods of reading in data see the Nextflow documentation on <a href=https://www.nextflow.io/docs/latest/channel.html#channel-factory>Channel Factories</a>.</p> <p>We can also do quite advanced things to manipuate data in channels, such as this:</p> <div class=highlight><pre><span></span><code><span class=n>Channel</span>
<span class=w>    </span><span class=o>.</span><span class=na>fromPath</span><span class=w> </span><span class=o>(</span><span class=w> </span><span class=n>params</span><span class=o>.</span><span class=na>metadata</span><span class=w> </span><span class=o>)</span>
<span class=w>    </span><span class=o>.</span><span class=na>splitCsv</span><span class=w> </span><span class=o>(</span><span class=w> </span><span class=nl>sep:</span><span class=w> </span><span class=s2>&quot;\t&quot;</span><span class=o>,</span><span class=w> </span><span class=nl>header:</span><span class=w> </span><span class=kc>true</span><span class=w> </span><span class=o>)</span>
<span class=w>    </span><span class=o>.</span><span class=na>map</span><span class=w>      </span><span class=o>{</span><span class=w> </span><span class=n>row</span><span class=w> </span><span class=o>-&gt;</span><span class=w> </span><span class=n>tuple</span><span class=o>(</span><span class=s2>&quot;${row.sample_id}&quot;</span><span class=o>,</span><span class=w> </span><span class=s2>&quot;${row.treatment}&quot;</span><span class=o>)</span><span class=w> </span><span class=o>}</span>
<span class=w>    </span><span class=o>.</span><span class=na>filter</span><span class=w>   </span><span class=o>{</span><span class=w> </span><span class=n>id</span><span class=o>,</span><span class=w> </span><span class=n>treatment</span><span class=w> </span><span class=o>-&gt;</span><span class=w> </span><span class=n>treatment</span><span class=w> </span><span class=o>!=</span><span class=w> </span><span class=s2>&quot;DMSO&quot;</span><span class=w> </span><span class=o>}</span>
<span class=w>    </span><span class=o>.</span><span class=na>unique</span><span class=w>   </span><span class=o>(</span><span class=w>  </span><span class=o>)</span>
<span class=w>    </span><span class=o>.</span><span class=na>set</span><span class=w>      </span><span class=o>{</span><span class=w> </span><span class=n>samples_and_treatments</span><span class=w> </span><span class=o>}</span>
</code></pre></div> <p>That's a bit of a handful! But what does it do? The first line specifies that we want to read some data from a file specified by the <code>metadata</code> parameter, and the second line actually reads that data using tab as delimiter, including a header. The <code>map</code> operator takes each entire row and subsets it to only two columns: the <code>sample_id</code> and <code>treatment</code> columns. This subset is stored as a tuple. The <code>filter</code> operator is then used to remove any tuples where the second entry, <code>treatment</code>, is not equal to the string <code>"DMSO"</code> (<em>i.e.</em> untreated cells, in this example). We then only take the unique tuples and set the results as the new channel <code>samples_and_treatments</code>. Let's say that this is the metadata we're reading:</p> <div class=highlight><pre><span></span><code>sample_id     dose    group     treatment
sample_1      0.1     control   DMSO
sample_1      1.0     control   DMSO
sample_1      2.0     control   DMSO
sample_2      0.1     case      vorinostat
sample_2      1.0     case      vorinostat
sample_2      2.0     case      vorinostat 
sample_3      0.1     case      fulvestrant
sample_3      1.0     case      fulvestrant
sample_3      2.0     case      fulvestrant
</code></pre></div> <p>Given the channel creation strategy above, we would get the following result:</p> <div class=highlight><pre><span></span><code>[sample_2, vorinostat]
[sample_3, fulvestrant]
</code></pre></div> <p>In this way, you can perform complex operations on input files or input metadata and send the resulting content to your downstream processes in a simple way. Composing data manipuations in Nextflow like this can be half the fun of writing the workflow. Check out Nextflow's documentation on <a href=https://www.nextflow.io/docs/latest/operator.html>Channel operators</a> to see the full list of channel operations at your disposal. </p> <h1 id=using-groovy-in-processes>Using Groovy in processes<a class=headerlink href=#using-groovy-in-processes title="Permanent link">#</a></h1> <p>You don't have to use bash or external scripts inside your processes all the time unless you want to: Nextflow is based on Groovy, which allows you to use both Groovy and Bash in the same process. For example, have a look at this:</p> <div class=highlight><pre><span></span><code><span class=n>process</span><span class=w> </span><span class=n>index_fasta</span><span class=w> </span><span class=o>{</span>
<span class=w>    </span><span class=n>tag</span><span class=w> </span><span class=s2>&quot;${fasta_name}&quot;</span>

<span class=w>    </span><span class=nl>input:</span>
<span class=w>    </span><span class=n>tuple</span><span class=w> </span><span class=n>val</span><span class=o>(</span><span class=n>fasta</span><span class=o>),</span><span class=w> </span><span class=n>path</span><span class=o>(</span><span class=n>fasta_file</span><span class=o>)</span>

<span class=w>    </span><span class=nl>output:</span>
<span class=w>    </span><span class=n>path</span><span class=o>(</span><span class=s2>&quot;${fasta_name}.idx&quot;</span><span class=o>),</span><span class=w> </span><span class=nl>emit:</span><span class=w> </span><span class=n>fasta</span>

<span class=w>    </span><span class=nl>script:</span>
<span class=w>    </span><span class=n>fasta_name</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=n>fasta</span><span class=o>.</span><span class=na>substring</span><span class=o>(</span><span class=mi>0</span><span class=o>,</span><span class=w> </span><span class=n>fasta</span><span class=o>.</span><span class=na>lastIndexOf</span><span class=o>(</span><span class=s2>&quot;.&quot;</span><span class=o>))</span>
<span class=w>    </span><span class=s2>&quot;&quot;&quot;</span>
<span class=s2>    index --ref ${fasta_file},${fasta_name}</span>
<span class=s2>    &quot;&quot;&quot;</span>
<span class=o>}</span>
</code></pre></div> <p>Here we have some command <code>index</code> that, for whatever reason, requires both the path to a FASTA file and the name of that file <em>without</em> the <code>.fasta</code> extension. We can use Groovy in the <code>script</code> directive together with normal Bash, mixing and matching as we like. The first line of the <code>script</code> directive gets the name of the FASTA file without the extension by removing anything after the dot, while the second calls the <code>index</code> command like normal using bash.</p> <h1 id=the-nf-core-pipeline-collection>The nf-core pipeline collection<a class=headerlink href=#the-nf-core-pipeline-collection title="Permanent link">#</a></h1> <p>You may have heard of the <a href=https://nf-co.re/ >nf-core</a> pipeline collection previously, which is a large, collaborative bioinformatics community dedicated to building, developing and maintaining Nextflow workflows. In fact, if you have sequenced data at <em>e.g.</em> the National Genomics Infrastructure (<a href=https://ngisweden.scilifelab.se/ >NGI</a>), you can be sure that the data processing has been run using one of the nf-core pipelines! While the community only started in 2018 (with a <a href=https://www.nature.com/articles/s41587-020-0439-x>Nature Biotechnology</a> paper in 2020), it already has over 30 production-ready pipelines with everything from genomics, transcriptomics, proteomics and metagenomics - and more being developed all the time.</p> <p>The nf-core pipelines all work in the same way, in that they have the same exact base for inputs, parameters and arguments, making them all highly similar to run. Since you've already learnt the basics of Nextflow in this course, you should now be able to also run the nf-core pipelines! It might be that you have a data type that you can analyse using one of the pipelines in nf-core, meaning you don't need to do anything other than find out what parameters you should run it with.</p> <p>Each pipeline comes with extensive documentation, test datasets that you can use to practice on, can be run on both HPCs like iTrop, cloud services like AWS or locally on your own computer. All pipelines support both Conda and Docker/Singularity, and you can additionally run specific versions of the pipelines, allowing for full reproducibility of your analyses. If you want to check nf-core out, simply head over to their <a href=https://nf-co.re/pipelines>list of pipelines</a> and see what's available! Who knows, you might even write your own nf-core pipeline in the future?</p> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../nextflow-6-optimising-the-mrsa-workflow/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Optmising the MRSA workflow" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <span class=md-footer__direction> Previous </span> <div class=md-ellipsis> Optmising the MRSA workflow </div> </div> </a> <a href=../../containers/containers-1-introduction/ class="md-footer__link md-footer__link--next" aria-label="Next: Introduction" rel=next> <div class=md-footer__title> <span class=md-footer__direction> Next </span> <div class=md-ellipsis> Introduction </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../../..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "header.autohide", "navigation.footer", "navigation.indexes", "navigation.tabs", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "toc.integrate"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../../../assets/javascripts/bundle.b4d07000.min.js></script> </body> </html>