<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="Workshop on Tools for Reproducible Research"><meta name=author content="Jacques Dainat et al."><link href=https://southgreenplatform.github.io/training_reproducible_research/pages/containers/others/docker/ rel=canonical><link rel=icon href=../../../../assets/images/favicon.png><meta name=generator content="mkdocs-1.4.3, mkdocs-material-9.1.15"><title>Introduction - Bioinformatics Workshop on Tools for Reproducible Research</title><link rel=stylesheet href=../../../../assets/stylesheets/main.26e3688c.min.css><link rel=stylesheet href=../../../../assets/stylesheets/palette.ecc896b0.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../../../stylesheets/extra.css><script>__md_scope=new URL("../../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#introduction class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-header__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../../images/southgreen.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Bioinformatics Workshop on Tools for Reproducible Research </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Introduction </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <a href=javascript:void(0) class="md-search__icon md-icon" title=Share aria-label=Share data-clipboard data-clipboard-text data-md-component=search-share tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg> </a> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../../../.. class=md-tabs__link> Home </a> </li> <li class=md-tabs__item> <a href=../../../introduction/introduction_southgreen/ class=md-tabs__link> Tutorials </a> </li> <li class=md-tabs__item> <a href=../../../cheat_sheet/bash/bash/ class=md-tabs__link> Cheat sheet </a> </li> <li class=md-tabs__item> <a href=../../../slido/ class=md-tabs__link> Ask a question </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../../../.. title="Bioinformatics Workshop on Tools for Reproducible Research" class="md-nav__button md-logo" aria-label="Bioinformatics Workshop on Tools for Reproducible Research" data-md-component=logo> <img src=../../../images/southgreen.png alt=logo> </a> Bioinformatics Workshop on Tools for Reproducible Research </label> <div class=md-nav__source> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> training_reproducible_research </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../.. class=md-nav__link> Home </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2> <label class=md-nav__link for=__nav_2 id=__nav_2_label tabindex=0> Tutorials <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Tutorials </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../introduction/introduction_southgreen/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../../introduction/roundtable/ class=md-nav__link> Roundtable Introduce Yourself </a> </li> <li class=md-nav__item> <a href=../../../introduction/introduction_rr/ class=md-nav__link> Introduction Reproducible Research </a> </li> <li class=md-nav__item> <a href=../../../data-management/data-management/ class=md-nav__link> Data Management </a> </li> <li class=md-nav__item> <a href=../../../ice-breaker/ class=md-nav__link> Ice breaker session </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_6> <label class=md-nav__link for=__nav_2_6 id=__nav_2_6_label tabindex=0> Git <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_6_label aria-expanded=false> <label class=md-nav__title for=__nav_2_6> <span class="md-nav__icon md-icon"></span> Git </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../git/git-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../../git/git-setup/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../../git/git-2-creating-repositories/ class=md-nav__link> Creating a git repository </a> </li> <li class=md-nav__item> <a href=../../../git/git-3-committing-changes/ class=md-nav__link> Committing changes </a> </li> <li class=md-nav__item> <a href=../../../git/git-undoing-and-recovering/ class=md-nav__link> Undoing and recovering </a> </li> <li class=md-nav__item> <a href=../../../git/git-4-ignoring-files/ class=md-nav__link> Ignoring files </a> </li> <li class=md-nav__item> <a href=../../../git/git-5-branches/ class=md-nav__link> Branching and merging </a> </li> <li class=md-nav__item> <a href=../../../git/git-6-tags/ class=md-nav__link> Tagging commits </a> </li> <li class=md-nav__item> <a href=../../../git/git-7-working-remotely/ class=md-nav__link> Working remotly </a> </li> <li class=md-nav__item> <a href=../../../git/git-8-conflicts/ class=md-nav__link> Conflict </a> </li> <li class=md-nav__item> <a href=../../../git/git-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_7> <label class=md-nav__link for=__nav_2_7 id=__nav_2_7_label tabindex=0> Conda <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_7_label aria-expanded=false> <label class=md-nav__title for=__nav_2_7> <span class="md-nav__icon md-icon"></span> Conda </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../conda/conda-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../../conda/conda-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../../conda/conda-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../../conda/conda-3-projects/ class=md-nav__link> Projects </a> </li> <li class=md-nav__item> <a href=../../../conda/conda-4-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8> <label class=md-nav__link for=__nav_2_8 id=__nav_2_8_label tabindex=0> Workflow Manager System <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_8_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8> <span class="md-nav__icon md-icon"></span> Workflow Manager System </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../wms/ class=md-nav__link> Introduction </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_2> <label class=md-nav__link for=__nav_2_8_2 id=__nav_2_8_2_label tabindex=0> Snakemake <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_2_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8_2> <span class="md-nav__icon md-icon"></span> Snakemake </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../snakemake/snakemake-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-3-visualising-workflows/ class=md-nav__link> Visualising workflow </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-4-the-mrsa-workflow/ class=md-nav__link> The MRSA Workflow </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-5-parameters/ class=md-nav__link> Parameters </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-6-logs/ class=md-nav__link> Logs </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-7-temporary-files/ class=md-nav__link> Temporary files </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-8-targets/ class=md-nav__link> Targets </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-9-shadow-rules/ class=md-nav__link> Shadow rules </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-10-generalizing-workflows/ class=md-nav__link> Generalizing workflows </a> </li> <li class=md-nav__item> <a href=../../../snakemake/snakemake-11-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_8_3> <label class=md-nav__link for=__nav_2_8_3 id=__nav_2_8_3_label tabindex=0> Nextflow <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_2_8_3_label aria-expanded=false> <label class=md-nav__title for=__nav_2_8_3> <span class="md-nav__icon md-icon"></span> Nextflow </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../nextflow/nextflow-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../../nextflow/nextflow-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../../nextflow/nextflow-nfcore/ class=md-nav__link> Getting started with nf-core </a> </li> <li class=md-nav__item> <a href=../../../nextflow/nextflow-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../../nextflow/nextflow-3-executing-workflows/ class=md-nav__link> Executing workflows </a> </li> <li class=md-nav__item> <a href=../../../nextflow/nextflow-4-working-with-processes/ class=md-nav__link> Working with process </a> </li> <li class=md-nav__item> <a href=../../../nextflow/nextflow-5-workflow-configuration/ class=md-nav__link> Workflow configuration </a> </li> <li class=md-nav__item> <a href=../../../nextflow/nextflow-6-optimising-the-mrsa-workflow/ class=md-nav__link> Optmising the MRSA workflow </a> </li> <li class=md-nav__item> <a href=../../../nextflow/nextflow-7-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_9> <label class=md-nav__link for=__nav_2_9 id=__nav_2_9_label tabindex=0> Containers <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_9_label aria-expanded=false> <label class=md-nav__title for=__nav_2_9> <span class="md-nav__icon md-icon"></span> Containers </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../containers-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../containers-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../containers-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../containers-3-building-images/ class=md-nav__link> Building a Docker image </a> </li> <li class=md-nav__item> <a href=../../containers-4-managing-containers/ class=md-nav__link> Managing containers </a> </li> <li class=md-nav__item> <a href=../../containers-5-sharing-images/ class=md-nav__link> Distributing your images </a> </li> <li class=md-nav__item> <a href=../../containers-6-packaging-the-case-study/ class=md-nav__link> Packaging the case study </a> </li> <li class=md-nav__item> <a href=../../containers-7-singularity/ class=md-nav__link> Singularity </a> </li> <li class=md-nav__item> <a href=../../containers-8-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_10> <label class=md-nav__link for=__nav_2_10 id=__nav_2_10_label tabindex=0> Notebook <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_10_label aria-expanded=false> <label class=md-nav__title for=__nav_2_10> <span class="md-nav__icon md-icon"></span> Notebook </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../notebook/notebook-context/ class=md-nav__link> Context </a> </li> <li class=md-nav__item> <a href=../../../notebook/notebook-markdown/ class=md-nav__link> Markdown </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_11> <label class=md-nav__link for=__nav_2_11 id=__nav_2_11_label tabindex=0> R Markdown <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_11_label aria-expanded=false> <label class=md-nav__title for=__nav_2_11> <span class="md-nav__icon md-icon"></span> R Markdown </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../rmarkdown/r-markdown-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../../rmarkdown/r-markdown-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../../rmarkdown/r-markdown-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../../rmarkdown/r-markdown-3-text/ class=md-nav__link> Text </a> </li> <li class=md-nav__item> <a href=../../../rmarkdown/r-markdown-4-code-chunks/ class=md-nav__link> Code chunks </a> </li> <li class=md-nav__item> <a href=../../../rmarkdown/r-markdown-5-the-yaml-header/ class=md-nav__link> The YAML header </a> </li> <li class=md-nav__item> <a href=../../../rmarkdown/r-markdown-6-rendering/ class=md-nav__link> Rendering </a> </li> <li class=md-nav__item> <a href=../../../rmarkdown/r-markdown-7-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../../rmarkdown/r-markdown-8-extra-material/ class=md-nav__link> Extra materiel </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_12> <label class=md-nav__link for=__nav_2_12 id=__nav_2_12_label tabindex=0> Jupyter <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_12_label aria-expanded=false> <label class=md-nav__title for=__nav_2_12> <span class="md-nav__icon md-icon"></span> Jupyter </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../jupyter/jupyter-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../../jupyter/jupyter-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../../jupyter/jupyter-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../../jupyter/jupyter-3-magics/ class=md-nav__link> Magics </a> </li> <li class=md-nav__item> <a href=../../../jupyter/jupyter-4-plotting/ class=md-nav__link> Plotting </a> </li> <li class=md-nav__item> <a href=../../../jupyter/jupyter-5-widgets/ class=md-nav__link> Widgets </a> </li> <li class=md-nav__item> <a href=../../../jupyter/jupyter-6-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../../../jupyter/jupyter-7-converting-notebooks/ class=md-nav__link> Converting notebooks </a> </li> <li class=md-nav__item> <a href=../../../jupyter/jupyter-8-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../../jupyter/jupyter-9-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_13> <label class=md-nav__link for=__nav_2_13 id=__nav_2_13_label tabindex=0> Quarto <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_13_label aria-expanded=false> <label class=md-nav__title for=__nav_2_13> <span class="md-nav__icon md-icon"></span> Quarto </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../quarto/quarto-1-introduction/ class=md-nav__link> Introduction </a> </li> <li class=md-nav__item> <a href=../../../quarto/quarto-installation/ class=md-nav__link> Setup </a> </li> <li class=md-nav__item> <a href=../../../quarto/quarto-2-the-basics/ class=md-nav__link> The basics </a> </li> <li class=md-nav__item> <a href=../../../quarto/quarto-3-the-rendering/ class=md-nav__link> Rendering </a> </li> <li class=md-nav__item> <a href=../../../quarto/quarto-4-the-extensions/ class=md-nav__link> Extensions </a> </li> <li class=md-nav__item> <a href=../../../quarto/quarto-5-the-mrsa-case-study/ class=md-nav__link> The MRSA case study </a> </li> <li class=md-nav__item> <a href=../../../quarto/quarto-6-comparison/ class=md-nav__link> Comparison </a> </li> <li class=md-nav__item> <a href=../../../quarto/quarto-7-extra-material/ class=md-nav__link> Extra material </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_2_14> <label class=md-nav__link for=__nav_2_14 id=__nav_2_14_label tabindex=0> Making a project reproducible <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_2_14_label aria-expanded=false> <label class=md-nav__title for=__nav_2_14> <span class="md-nav__icon md-icon"></span> Making a project reproducible </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../project/project-1-review/ class=md-nav__link> Review </a> </li> <li class=md-nav__item> <a href=../../../project/project-2-good_praticies/ class=md-nav__link> Good praticies </a> </li> <li class=md-nav__item> <a href=../../../project/project-3-and_now/ class=md-nav__link> And now ? </a> </li> <li class=md-nav__item> <a href=../../../project/project-4-demo/ class=md-nav__link> To end on a high note ! </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../../take_down/take_down/ class=md-nav__link> Take down </a> </li> <li class=md-nav__item> <a href=../../../final_word/ class=md-nav__link> Final word </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3 id=__nav_3_label tabindex=0> Cheat sheet <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Cheat sheet </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../cheat_sheet/bash/bash/ class=md-nav__link> Bash </a> </li> <li class=md-nav__item> <a href=../../../cheat_sheet/git/git/ class=md-nav__link> Git </a> </li> <li class=md-nav__item> <a href=../../../cheat_sheet/git/git_official/ class=md-nav__link> Git official </a> </li> <li class=md-nav__item> <a href=../../../cheat_sheet/conda/conda/ class=md-nav__link> Conda </a> </li> <li class=md-nav__item> <a href=../../../cheat_sheet/notebook/notebook/ class=md-nav__link> Notebook </a> </li> <li class=md-nav__item> <a href=https://dockerlabs.collabnix.com/docker/cheatsheet/ class=md-nav__link> Docker </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../../slido/ class=md-nav__link> Ask a question </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/edit/main/docs/pages/containers/others/docker.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg> </a> <a href=https://github.com/SouthGreenPlatform/training_reproducible_research/raw/main/docs/pages/containers/others/docker.md title="View source of this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.15 8.15 0 0 1-1.23-2Z"/></svg> </a> <h1 id=introduction>Introduction<a class=headerlink href=#introduction title="Permanent link">#</a></h1> <p>Docker is a tool designed to make it easier to create, deploy, and run applications by isolating them in "containers". The idea is to package your code together with everything it needs (other packages it depends on, various environment settings, data..) into one unit, <em>i.e.</em> a container. This way we can ensure that the code results in exactly the same output regardless of where it's executed. Containers are in many ways similar to virtual machines but more lightweight. Rather than starting up a whole new OS, Docker containers can use the same Linux kernel as the system that they're running on. This makes them much faster and smaller compared to virtual machines. While this might sound a bit technical, actually using Docker is quite easy, fun and very powerful.</p> <p>Just as with Git, Docker was designed for software development but is rapidly becoming widely used in scientific research. Say that you are building a web application. You could then run the web server in one container and the database in another, thereby reducing the risk of one system affecting the other in unpredictable ways. Docker containers have also proven to be a very good solution for packaging, running and distributing scientific data analyses. Some applications relevant for reproducible research can be:</p> <ul> <li>When publishing, package your whole analysis pipeline together with your data in a Docker image and let it accompany the article. This way interested readers can reproduce your analysis at the push of a button.</li> <li>Packaging your analysis in a Docker container enables you to develop on <em>e.g.</em> your laptop and then seamlessly move to cluster or cloud to run the actual analysis.</li> <li>Say that you are collaborating on a project and you are using Mac while your collaborator is using Windows. You can then set up a Docker image specific for your project to ensure that you are working in an identical environment.</li> </ul> <p>All of this might sound a bit abstract so far, but it'll become more clear during the exercises. If you want to read more you can check out these resources:</p> <ul> <li>A "Get started with Docker" at <a href=https://docs.docker.com/get-started/ >docker.com</a>.</li> <li>An <a href=https://arxiv.org/abs/1410.0846>early paper</a> on the subject of using Docker for reproducible research.</li> </ul> <p>This tutorial depends on files from the course GitHub repo. Take a look at the <a href=setup.md>setup</a> for instructions on how to set it up if you haven't done so already. Then open up a terminal and go to <code>workshop-reproducible-research/docker</code>.</p> <div class="admonition attention"> <p class=admonition-title>Attention</p> <p>Docker images tend to take up quite a lot of space. In order to do all the exercises in this tutorial you need to have ~10 GB available.</p> </div> <h2 id=the-basics>The basics<a class=headerlink href=#the-basics title="Permanent link">#</a></h2> <p>We're almost ready to start, just one last note on nomenclature. You might have noticed that we sometimes refer to "Docker images" and sometimes to "Docker containers". A container is simply an instance of an image. To use a programming metaphor, if an image is a class, then a container is an instance of that class — a runtime object. You can have an image containing, say, a certain Linux distribution, and then start multiple containers running that same OS.</p> <div class="admonition attention"> <p class=admonition-title>Attention</p> <p>If you don't have root privileges you have to prepend all Docker commands with <code>sudo</code> (see the tip above)</p> </div> <p>Docker containers typically run Linux, so let's start by downloading an image containing Ubuntu (a popular Linux distribution that is based on only open-source tools) through the command line.</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>pull<span class=w> </span>ubuntu:latest
</code></pre></div> <p>You will notice that it downloads different layers with weird hashes as names. This represents a very fundamental property of Docker images that we'll get back to in just a little while. The process should end with something along the lines of:</p> <div class=highlight><pre><span></span><code>Status: Downloaded newer image for ubuntu:latest
docker.io/library/ubuntu:latest
</code></pre></div> <p>Let's take a look at our new and growing collection of Docker images:</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>images
</code></pre></div> <p>The Ubuntu image show show up in this list, with something looking like this:</p> <div class=highlight><pre><span></span><code>REPOSITORY       TAG              IMAGE ID            CREATED             SIZE
ubuntu           latest           d70eaf7277ea        3 weeks ago         72.9MB
</code></pre></div> <p>We can now start a container running our image. We can refer to the image either by "REPOSITORY:TAG" ("latest" is the default so we can omit it) or "IMAGE ID". The syntax for <code>docker run</code> is <code>docker run [OPTIONS] IMAGE [COMMAND] [ARG...]</code>. Let's run the command <code>uname -a</code> to get some info about the operating system. First run on your own system (skip this if you're using Windows via the Windows 10 PowerShell, or use <code>systeminfo</code> which is the Windows equivalent):</p> <div class=highlight><pre><span></span><code>uname<span class=w> </span>-a
</code></pre></div> <p>This should print something like this to your command line:</p> <div class=highlight><pre><span></span><code>Darwin liv433l.lan 15.6.0 Darwin Kernel Version 15.6.0: Mon Oct  2 22:20:08 PDT 2017; root:xnu-3248.71.4~1/RELEASE_X86_64 x86_64
</code></pre></div> <p>Seems like I'm running the Darwin version of macOS. Then run it in the Ubuntu Docker container:</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>run<span class=w> </span>ubuntu<span class=w> </span>uname<span class=w> </span>-a
</code></pre></div> <p>Here I get the following result:</p> <div class=highlight><pre><span></span><code>Linux 24d063b5d877 5.4.39-linuxkit #1 SMP Fri May 8 23:03:06 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
</code></pre></div> <p>And now I'm running on Linux! Try the same thing with <code>whoami</code>.</p> <p>So, seems we can execute arbitrary commands on Linux. Seems useful, but maybe a bit limited. We can also get an interactive terminal with the flags <code>-it</code>.</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>run<span class=w> </span>-it<span class=w> </span>ubuntu
</code></pre></div> <p>This should put at a terminal prompt inside a container running Ubuntu. Your prompt should now look similar to:</p> <div class=highlight><pre><span></span><code>root@1f339e929fa9:/#
</code></pre></div> <p>Here you can do whatever; install, run, remove stuff. It will still be within the container and never affect your host system. Now exit the container with <code>exit</code>.</p> <p>Ok, so Docker lets us work in any OS in a quite convenient way. That would probably be useful on its own, but Docker is much more powerful than that.</p> <div class="admonition note"> <p class=admonition-title>Quick recap</p> <p>In this section we've learned:</p> <ul> <li>How to use <code>docker pull</code> for downloading images from a central registry.</li> <li>How to use <code>docker images</code> for getting information about the images we have on our system.</li> <li>How to use <code>docker run</code> for starting a container from an image.</li> <li>How to use the <code>-it</code> flag for running in interactive mode.</li> </ul> </div> <h2 id=building-a-docker-image>Building a Docker image<a class=headerlink href=#building-a-docker-image title="Permanent link">#</a></h2> <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script> <p>In the previous section we downloaded a Docker image of Ubuntu and noticed that it was based on layers, each with a unique hash as id. An image in Docker is based on a number of read-only layers, where each layer contains the differences to the previous layers. If you've done the <a href=git.md>Git tutorial</a> this might remind you of how a Git commit contains the difference to the previous commit. The great thing about this is that we can start from one base layer, say containing an operating system and some utility programs, and then generate many new images based on this, say 10 different project-specific images. The total space requirements would then only be $base+\sum_{i=1}^{10}(specific_{i})$ rather than $\sum_{i=1}^{10}(base+specific_{i})$. For example, Bioconda (see the <a href=conda.md>Conda tutorial</a>) has one base image and then one individual layer for each of the &gt;3000 packages available in Bioconda.</p> <p>Docker provides a convenient way to describe how to go from a base image to the image we want by using a "Dockerfile". This is a simple text file containing the instructions for how to generate each layer. Docker images are typically quite large, often several GBs, while Dockerfiles are small and serve as blueprints for the images. It is therefore good practice to have your Dockerfile in your project Git repository, since it allows other users to exactly replicate your project environment.</p> <p>If you've been doing these tutorials on Windows you've been using the Docker image <code>nbisweden/workshop-reproducible-research:slim</code>. The Dockerfile for generating that image is called <code>Dockerfile_slim</code> and is located in your <code>docker</code> directory (where you should hopefully be standing already). We will now go through that file and discuss the different steps and what they do. After that we'll build the image and test it out. Lastly, we'll start from that image and make a new one to reproduce the results from the <a href=conda.md>Conda tutorial</a>.</p> <p>Here are the first few lines of <code>Dockerfile_slim</code>. Each line in the Dockerfile will typically result in one layer in the resulting image. The format for Dockerfiles is <code>INSTRUCTION arguments</code>. A full specification of the format, together with best practices, can be found <a href=https://docs.docker.com/engine/reference/builder/ >here</a>.</p> <div class=highlight><pre><span></span><code>FROM ubuntu:16.04

LABEL description = &quot;Minimal image for the NBIS reproducible research course.&quot;
MAINTAINER &quot;John Sundh&quot; john.sundh@scilifelab.se
</code></pre></div> <p>Here we use the instructions <code>FROM</code>, <code>LABEL</code> and <code>MAINTAINER</code>. The important one is <code>FROM</code>, which specifies the base image our image should start from. In this case we want it to be Ubuntu 16.04, which is one of the <a href=https://hub.docker.com/explore/ >official repositories</a>. There are many roads to Rome when it comes to choosing the best image to start from. Say you want to run RStudio in a Conda environment through a Jupyter notebook. You could then start from one of the <a href=https://github.com/rocker-org/rocker>rocker images</a> for R, a <a href=https://hub.docker.com/r/continuumio/miniconda/ >Miniconda image</a>, or a <a href=https://hub.docker.com/r/jupyter/ >Jupyter image</a>. Or you just start from one of the low-level official images and set up everything from scratch. <code>LABEL</code> and <code>MAINTAINER</code> is just meta-data that can be used for organizing your various Docker components.</p> <p>Let's take a look at the next section of the Dockerfile.</p> <div class=highlight><pre><span></span><code># Use bash as shell
SHELL [&quot;/bin/bash&quot;, &quot;-c&quot;]

# Set workdir
WORKDIR /course
</code></pre></div> <p><code>SHELL</code> simply sets which shell to use. <code>WORKDIR</code> determines the directory the container should start in. The next few lines introduce the important <code>RUN</code> instruction, which is used for executing shell commands:</p> <div class=highlight><pre><span></span><code># Install necessary tools
RUN apt-get update &amp;&amp; \
    apt-get install -y --no-install-recommends bzip2 \
                                               ca-certificates \
                                               curl \
                                               fontconfig \
                                               git \
                                               language-pack-en \
                                               tzdata \
                                               vim \
                                               unzip \
                                               wget \
    &amp;&amp; apt-get clean

# Install Miniconda and add to PATH
RUN curl -L https://repo.continuum.io/miniconda/Miniconda3-4.7.12.1-Linux-x86_64.sh -O &amp;&amp; \
    bash Miniconda3-4.7.12.1-Linux-x86_64.sh -bf -p /usr/miniconda3/ &amp;&amp; \
    rm Miniconda3-4.7.12.1-Linux-x86_64.sh &amp;&amp; \
    /usr/miniconda3/bin/conda clean -tipsy &amp;&amp; \
    ln -s /usr/miniconda3/etc/profile.d/conda.sh /etc/profile.d/conda.sh &amp;&amp; \
    echo &quot;. /usr/miniconda3/etc/profile.d/conda.sh&quot; &gt;&gt; ~/.bashrc &amp;&amp; \
    echo &quot;conda activate base&quot; &gt;&gt; ~/.bashrc
</code></pre></div> <p>As a general rule, you want each layer in an image to be a "logical unit". For example, if you want to install a program the <code>RUN</code> command should both retrieve the program, install it and perform any necessary clean up. This is due to how layers work and how Docker decides what needs to be rerun between builds. The first command uses Ubuntu's package manager APT to install some packages (similar to how we've previously used Conda). Say that the first command was split into two instead:</p> <div class=highlight><pre><span></span><code># Update apt-get
RUN apt-get update

# Install packages
RUN apt-get install -y --no-install-recommends bzip2 \
                                               ca-certificates \
                                               curl \
                                               fontconfig \
                                               git \
                                               language-pack-en \
                                               tzdata \
                                               vim \
                                               unzip \
                                               wget

# Clear the local repository of retrieved package files
RUN apt-get clean
</code></pre></div> <p>The first command will update the apt-get package lists and the second will install the packages <code>bzip2</code>, <code>ca-certificates</code>, <code>curl</code>, <code>fontconfig</code>, <code>git</code>, <code>language-pack-en</code>, <code>tzdata</code>, <code>vim</code>, <code>unzip</code> and <code>wget</code>. Say that you build this image now, and then in a month's time you realize that you would have liked a Swedish language pack instead of an English. You change to <code>language-pack-sv</code> and rebuild the image. Docker detects that there is no layer with the new list of packages and reruns the second <code>RUN</code> command. <strong>However, there is no way for Docker to know that it should also update the apt-get package lists</strong>. You therefore risk to end up with old versions of packages and, even worse, the versions would depend on when the previous version of the image was first built.</p> <p>The next <code>RUN</code> command retrieves and installs Miniconda3. Let's see what would happen if we had that as separate commands instead.</p> <div class=highlight><pre><span></span><code># Download Miniconda3
RUN curl -L https://repo.continuum.io/miniconda/Miniconda3-4.7.12.1-Linux-x86_64.sh -O

# Install it
RUN bash Miniconda3-4.7.12.1-Linux-x86_64.sh -bf -p /usr/miniconda3/

# Remove the downloaded installation file
RUN rm Miniconda3-4.7.12.1-Linux-x86_64.sh

# Remove unused packages and caches
RUN /usr/miniconda3/bin/conda clean -tipsy

# Permanently enable the Conda command
RUN ln -s /usr/miniconda3/etc/profile.d/conda.sh /etc/profile.d/conda.sh
RUN echo &quot;. /usr/miniconda3/etc/profile.d/conda.sh&quot; &gt;&gt; ~/.bashrc

# Add the base environment permanently to PATH
RUN echo &quot;conda activate base&quot; &gt;&gt; ~/.bashrc
</code></pre></div> <p>Remember that each layer contains the difference compared to the previous layer? What will happen here is that the first command adds the installation file and the second will unpack the file and install the software. The third layer will say "the installation file should no longer exist on the file system". However, the file will still remain in the image since the image is constructed layer-by-layer bottom-up. This results in unnecessarily many layers and bloated images. Line four is cleaning up conda to free up space, and the next two lines are there to make the Conda command available in the shell. The last command adds a code snippet to the bash startup file which automatically activates the Conda base environment in the container.</p> <div class=highlight><pre><span></span><code># Add conda to PATH and set locale
ENV PATH=&quot;/usr/miniconda3/bin:${PATH}&quot;
ENV LC_ALL en_US.UTF-8
ENV LC_LANG en_US.UTF-8
</code></pre></div> <p>Here we use the new instruction <code>ENV</code>. The first command adds <code>conda</code> to the path, so we can write <code>conda install</code> instead of <code>/usr/miniconda3/bin/conda install</code>. The next two commands set an UTF-8 character encoding so that we can use weird characters (and a bunch of other things).</p> <div class=highlight><pre><span></span><code># Configure Conda channels and install Mamba
RUN conda config --add channels bioconda \
    &amp;&amp; conda config --add channels conda-forge \
    &amp;&amp; conda config --set channel_priority strict \
    &amp;&amp; conda install mamba \
    &amp;&amp; mamba clean --all
</code></pre></div> <p>Here we just configure Conda and install Mamba, for quicker installations of any subsequent Conda packages we might want to do.</p> <div class=highlight><pre><span></span><code># Open port for running Jupyter Notebook
EXPOSE 8888

# Start Bash shell by default
CMD /bin/bash
</code></pre></div> <p><code>EXPOSE</code> opens up the port 8888, so that we can later run a Jupyter Notebook server on that port. <code>CMD</code> is an interesting instruction. It sets what a container should run when nothing else is specified. It can be used for example for printing some information on how to use the image or, as here, start a shell for the user. If the purpose of your image is to accompany a publication then <code>CMD</code> could be to run the workflow that generates the paper figures from raw data.</p> <p>Ok, so now we understand how a Dockerfile works. Constructing the image from the Dockerfile is really simple. Try it out now:</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>build<span class=w> </span>-f<span class=w> </span>Dockerfile_slim<span class=w> </span>-t<span class=w> </span>my_docker_image<span class=w> </span>.
</code></pre></div> <p>This should result in something similar to this:</p> <div class=highlight><pre><span></span><code>=&gt; [internal] load build definition from Dockerfile_slim
=&gt; =&gt; transferring dockerfile: 1.88kB
=&gt; [internal] load .dockerignore
=&gt; =&gt; transferring context: 2B
=&gt; [internal] load metadata for docker.io/library/ubuntu:16.04
=&gt; [auth] library/ubuntu:pull token for registry-1.docker.io
=&gt; [1/5] FROM docker.io/library/ubuntu:16.04@sha256:bb84bbf2ff36d46acaf0bb0c6bcb33dae64cd93cba8652d74c9aaf438fada438
=&gt; CACHED [2/5] WORKDIR /course
=&gt; CACHED [3/5] RUN apt-get update &amp;&amp;     apt-get install -y --no-install-recommends bzip2                                                ca-certificates
=&gt; CACHED [4/5] RUN curl -L https://repo.continuum.io/miniconda/Miniconda3-4.7.12.1-Linux-x86_64.sh -O &amp;&amp;   h Miniconda3-4.7.12.1-Linux-x86_64.sh -bf -p /usr/miniconda3
=&gt; CACHED [5/5] RUN conda config --add channels bioconda     &amp;&amp; conda config --add channels conda-forge     &amp;&amp; conda config --set channel_priority strict     &amp;&amp; conda instal
=&gt; exporting to image
=&gt; =&gt; exporting layers
=&gt; =&gt; writing image sha256:d14301f829d4554816df54ace927ec0aaad4a994e028371455f7a18a370f6af9
=&gt; =&gt; naming to docker.io/library/my_docker_image
</code></pre></div> <p>Exactly how the output looks depends on which version of Docker you are using. The <code>-f</code> flag sets which Dockerfile to use and <code>-t</code> tags the image with a name. This name is how you will refer to the image later. Lastly, the <code>.</code> is the path to where the image should be build (<code>.</code> means the current directory). This had no real impact in this case, but matters if you want to import files. Validate with <code>docker images</code> that you can see your new image.</p> <p>Now it's time to make our own Dockerfile to reproduce the results from the <a href=conda>Conda tutorial</a>. If you haven't done the tutorial, it boils down to creating a Conda environment file, setting up that environment, downloading three RNA-seq data files, and running FastQC on those files. We will later package and run the whole RNA-seq workflow in a Docker container, but for now we keep it simple to reduce the size and time required.</p> <p>The Conda tutorial uses a shell script, <code>run_qc.sh</code>, for downloading and running the analysis. A copy of this file should also be available in your current directory. If we want to use the same script we need to include it in the image. So, this is what we need to do:</p> <ol> <li> <p>Create the file <code>Dockerfile_conda</code>.</p> </li> <li> <p>Set <code>FROM</code> to the image we just built.</p> </li> <li> <p>Install the required packages with Conda. We could do this by adding <code>environment.yml</code> from the Conda tutorial, but here we do it directly as <code>RUN</code> commands. We need to add the conda-forge and bioconda channels with <code>conda config --add channels &lt;channel_name&gt;</code> and install <code>fastqc=0.11.9</code> and <code>sra-tools=2.10.1</code> with <code>conda install</code>. The packages will be installed to the default environment named <code>base</code> inside the container.</p> </li> <li> <p>Add <code>run_qc.sh</code> to the image by using the <code>COPY</code> instruction. The syntax is <code>COPY source target</code>, so in our case simply <code>COPY run_qc.sh .</code> to copy to the work directory in the image.</p> </li> <li> <p>Set the default command for the image to <code>bash run_qc.sh</code>, which will execute the shell script.</p> </li> </ol> <p>Try to add required lines to <code>Dockerfile_conda</code>. If it seems overwhelming you can take a look below</p> <details class=example> <summary>Click to show an example of <code>Dockerfile_conda</code></summary> <div class=highlight><pre><span></span><code>FROM my_docker_image:latest
RUN conda config --add channels bioconda &amp;&amp; \
    conda config --add channels conda-forge &amp;&amp; \
    conda install -n base fastqc=0.11.9 sra-tools=2.10.1
COPY run_qc.sh .
CMD bash run_qc.sh
</code></pre></div> </details> <p>Build the image and tag it <code>my_docker_conda</code>:</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>build<span class=w> </span>-t<span class=w> </span>my_docker_conda<span class=w> </span>-f<span class=w> </span>Dockerfile_conda<span class=w> </span>.
</code></pre></div> <p>Verify that the image was built using <code>docker images</code>.</p> <div class="admonition note"> <p class=admonition-title>Quick recap</p> <p>In this section we've learned:</p> <ul> <li>How the keywords <code>FROM</code>, <code>LABEL</code>, <code>MAINTAINER</code>, <code>RUN</code>, <code>ENV</code>, <code>SHELL</code>, <code>WORKDIR</code>, and <code>CMD</code> can be used when writing a Dockerfile.</li> <li>The importance of letting each layer in the Dockerfile be a "logical unit".</li> <li>How to use <code>docker build</code> to construct and tag an image from a Dockerfile.</li> </ul> </div> <h2 id=managing-containers>Managing containers<a class=headerlink href=#managing-containers title="Permanent link">#</a></h2> <p>When you start a container with <code>docker run</code> it is given an unique id that you can use for interacting with the container. Let's try to run a container from the image we just created:</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>run<span class=w> </span>my_docker_conda
</code></pre></div> <p>If everything worked <code>run_qc.sh</code> is executed and will first download and then analyse the three samples. Once it's finished you can list all containers, including those that have exited.</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>container<span class=w> </span>ls<span class=w> </span>--all
</code></pre></div> <p>This should show information about the container that we just ran. Similar to: <div class=highlight><pre><span></span><code>CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS                      PORTS               NAMES
39548f30ce45        my_docker_conda     &quot;/bin/bash -c &#39;bas...&quot;    3 minutes ago       Exited (0) 3 minutes ago                             el
</code></pre></div></p> <p>If we run <code>docker run</code> without any flags, your local terminal is attached to the container. This enables you to see the output of <code>run_qc.sh</code>, but also disables you from doing anything else in the meantime. We can start a container in detached mode with the <code>-d</code> flag. Try this out and run <code>docker container ls</code> to validate that the container is running.</p> <p>By default, Docker keeps containers after they have exited. This can be convenient for debugging or if you want to look at logs, but it also consumes huge amounts of disk space. It's therefore a good idea to always run with <code>--rm</code>, which will remove the container once it has exited.</p> <p>If we want to enter a running container, there are two related commands we can use, <code>docker attach</code> and <code>docker exec</code>. <code>docker attach</code> will attach local standard input, output, and error streams to a running container. This can be useful if your terminal closed down for some reason or if you started a terminal in detached mode and changed your mind. <code>docker exec</code> can be used to execute any command in a running container. It's typically used to peak in at what is happening by opening up a new shell. Here we start the container in detached mode and then start a new interactive shell so that we can see what happens. If you use <code>ls</code> inside the container you can see how the script generates file in the <code>data</code>, <code>intermediate</code> and <code>results</code> directories. Note that you will be thrown out when the container exits, so you have to be quick.</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>run<span class=w> </span>-d<span class=w> </span>--rm<span class=w> </span>--name<span class=w> </span>my_container<span class=w> </span>my_docker_conda
docker<span class=w> </span><span class=nb>exec</span><span class=w> </span>-it<span class=w> </span>my_container<span class=w> </span>/bin/bash
</code></pre></div> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>Sometimes you would like to enter a stopped container. It's not a common use case, but it's included here for those of you who are doing these tutorials on Windows using Docker. Inadvertently shutting down your container can result in loss of a lot of work if you're not able to restart it. If you were to use <code>docker start</code> it would rerun the command set by <code>CMD</code>, which may not be what you want. Instead we use <code>docker commit container_name new_image_name</code> to convert the container <code>container_name</code> to the image <code>new_image_name</code>. We can then start a new container in that image as we normally would with <code>docker run -it --rm new_image_name /bin/bash</code>. Confusing, right? In theory, this would allow you to bypass using Dockerfiles and instead generate your image by entering an empty container in interactive mode, install everything there, and then commit as a new image. However, by doing this you would lose many of the advantages that Dockerfiles provide, such as easy distribution and efficient space usage via layers.</p> </div> <div class="admonition note"> <p class=admonition-title>Quick recap</p> <p>In this section we've learned:</p> <ul> <li>How to use <code>docker run</code> for starting a container and how the flags <code>-d</code> and <code>--rm</code> work.</li> <li>How to use <code>docker container ls</code> for displaying information about the containers.</li> <li>How to use <code>docker attach</code> and <code>docker exec</code> to interact with running containers.</li> </ul> </div> <h3 id=bind-mounts>Bind mounts<a class=headerlink href=#bind-mounts title="Permanent link">#</a></h3> <p>There are obviously some advantages to isolating and running your data analysis in containers, but at some point you need to be able to interact with the host system to actually deliver the results. This is done via bind mounts. When you use a bind mount, a file or directory on the <em>host machine</em> is mounted into a container. That way, when the container generates a file in such a directory it will appear in the mounted directory on your host system.</p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>Docker also has a more advanced way of data storage called <a href=https://docs.docker.com/engine/admin/volumes/ >volumes</a>. Volumes provide added flexibility and are independent of the host machine's filesystem having a specific directory structure available. They are particularly useful when you want to share data <em>between</em> containers.</p> </div> <p>Say that we are interested in getting the resulting html reports from FastQC in our container. We can do this by mounting a directory called, say, <code>fastqc_results</code> in your current directory to the <code>/course/results/fastqc</code> directory in the container. Try this out by running:</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>run<span class=w> </span>--rm<span class=w> </span>-v<span class=w> </span><span class=k>$(</span><span class=nb>pwd</span><span class=k>)</span>/fastqc_results:/course/results/fastqc<span class=w> </span>my_docker_conda
</code></pre></div> <p>Here the <code>-v</code> flag to docker run specifies the bind mount in the form of <code>directory/on/your/computer:/directory/inside/container</code>. <code>$(pwd)</code> simply evaluates to the working directory on your computer.</p> <p>Once the container finishes validate that it worked by opening one of the html reports under <code>fastqc_results/</code>.</p> <p>We can also use bind mounts for getting files into the container rather than out. We've mainly been discussing Docker in the context of packaging an analysis pipeline to allow someone else to reproduce its outcome. Another application is as a kind of very powerful environment manager, similarly to how we've used Conda before. If you've organized your work into projects, then you can mount the whole project directory in a container and use the container as the terminal for running stuff while still using your normal OS for editing files and so on. Let's try this out by mounting our current directory and start an interactive terminal. Note that this will override the <code>CMD</code> command, so we won't start the analysis automatically when we start the container.</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>run<span class=w> </span>-it<span class=w> </span>--rm<span class=w> </span>-v<span class=w> </span><span class=k>$(</span><span class=nb>pwd</span><span class=k>)</span>:/course/<span class=w> </span>my_docker_conda<span class=w> </span>/bin/bash
</code></pre></div> <p>If you run <code>ls</code> you will see that all the files in the <code>docker</code> directory are there. Now edit <code>run_qc.sh</code> <strong>on your host system</strong> to download, say, 12000 reads instead of 15000. Then rerun the analysis with <code>bash run_qc.sh</code>. Tada! Validate that the resulting html reports look fine and then exit the container with <code>exit</code>.</p> <h2 id=distributing-your-images>Distributing your images<a class=headerlink href=#distributing-your-images title="Permanent link">#</a></h2> <p>There would be little point in going through all the trouble of making your analyses reproducible if you can't distribute them to others. Luckily, sharing Docker containers is extremely easy. The most common way is to use <a href=https://hub.docker.com>Dockerhub</a>. Dockerhub lets you host unlimited public images and one private image for free, after that they charge a small fee. If you want to try it out here is how to do it:</p> <ol> <li> <p>Register for an account on <a href=https://hub.docker.com>Dockerhub</a>.</p> </li> <li> <p>Use <code>docker login -u your_dockerhub_id</code> to login to the Dockerhub registry.</p> </li> <li> <p>When you build an image, tag it with <code>-t your_dockerhub_id/image_name</code>, rather than just <code>image_name</code>.</p> </li> <li> <p>Once the image has been built, upload it to Dockerhub with <code>docker push your_dockerhub_id/image_name</code>.</p> </li> <li> <p>If another user runs <code>docker run your_dockerhub_id/image_name</code> the image will automatically be retrieved from Dockerhub. You can use <code>docker pull</code> for downloading without running.</p> </li> </ol> <p>If you want to refer to a Docker image in for example a publication, it's very important that it's the correct version of the image. You can do this by adding a tag to the name like this <code>docker build -t your_dockerhub_id/image_name:tag_name</code>.</p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>On Dockerhub it is also possible to link to your Bitbucket or GitHub account and select repositories from which you want to automatically build and distribute Docker images. The Dockerhub servers will then build an image from the Dockerfile in your repository and make it available for download using <code>docker pull</code>. That way, you don't have to bother manually building and pushing using <code>docker push</code>. The GitHub repository for this course is linked to Dockerhub and the Docker images are built automatically from <code>Dockerfile</code> and <code>Dockerfile_slim</code>, triggered by changes made to the GitHub repository. You can take a look at the course on Dockerhub <a href=https://hub.docker.com/r/nbisweden/workshop-reproducible-research>here</a>.</p> </div> <h2 id=packaging-the-case-study>Packaging the case study<a class=headerlink href=#packaging-the-case-study title="Permanent link">#</a></h2> <p>During these tutorials we have been working on a case study about the multiresistant bacteria MRSA. Here we will build and run a Docker container that contains all the work we've done so far.</p> <ul> <li>We've <a href=git.md>set up a GitHub repository</a> for version control and for hosting our project.</li> <li>We've defined a <a href=conda.md>Conda environment</a> that specifies the packages we're depending on in the project.</li> <li>We've constructed a <a href=snakemake.md>Snakemake workflow</a> that performs the data analysis and keeps track of files and parameters.</li> <li>We've written a <a href=rmarkdown.md>R Markdown document</a> that takes the results from the Snakemake workflow and summarizes them in a report.</li> </ul> <p>The <code>docker</code> directory contains the final versions of all the files we've generated in the other tutorials: <code>environment.yml</code>, <code>Snakefile</code>, <code>config.yml</code>, <code>code/header.tex</code>, and <code>code/supplementary_material.Rmd</code>. The only difference compared to the other tutorials is that we have also included the rendering of the Supplementary Material HTML file into the Snakemake workflow as the rule <code>make_supplementary</code>. Running all of these steps will take some time to execute (around 20 minutes or so), in particular if you're on a slow internet connection, and result in a 3.75 GB image.</p> <p>Now take a look at <code>Dockerfile</code>. Everything should look quite familiar to you, since it's basically the same steps as in the image we constructed in the previous section, although some sections have been moved around. The main difference is that we add the project files needed for executing the workflow (mentioned in the previous paragraph), and install the conda packages listed in <code>environment.yml</code>. If you look at the <code>CMD</code> command you can see that it will run the whole Snakemake workflow by default.</p> <p>Now run <code>docker build</code> as before, tag the image with <code>my_docker_project</code>:</p> <p><div class=highlight><pre><span></span><code>docker<span class=w> </span>build<span class=w> </span>-t<span class=w> </span>my_docker_project<span class=w> </span>-f<span class=w> </span>Dockerfile<span class=w> </span>.
</code></pre></div> and go get a coffee while the image builds (or you could use <code>docker pull nbisweden/workshop-reproducible-research</code> which will download the same image).</p> <p>Validate with <code>docker images</code>. Now all that remains is to run the whole thing with <code>docker run</code>. We just want to get the results, so mount the directory <code>/course/results/</code> to, say, <code>mrsa_results</code> in your current directory.</p> <p>Well done! You now have an image that allows anyone to exactly reproduce your analysis workflow (if you first <code>docker push</code> to Dockerhub that is).</p> <div class="admonition tip"> <p class=admonition-title>Tip</p> <p>If you've done the <a href=jupyter.md>Jupyter Notebook tutorial</a>, you know that Jupyter Notebook runs as a web server. This makes it very well suited for running in a Docker container, since we can just expose the port Jupyter Notebook uses and redirect it to one of our own. You can then work with the notebooks in your browser just as you've done before, while it's actually running in the container. This means you could package your data, scripts and environment in a Docker image that also runs a Jupyter Notebook server. If you make this image available, say on Dockerhub, other researchers could then download it and interact with your data/code via the fancy interactive Jupyter notebooks that you have prepared for them. We haven't made any fancy notebooks for you, but we <em>have</em> set up a Jupyter Notebook server. Try it out if you want to (replace the image name with your version if you've built it yourself):</p> <div class=highlight><pre><span></span><code>docker<span class=w> </span>run<span class=w> </span>-it<span class=w> </span>-p<span class=w> </span><span class=m>8888</span>:8888<span class=w> </span>nbisweden/workshop-reproducible-research<span class=w> </span><span class=se>\</span>
<span class=w>    </span>jupyter<span class=w> </span>notebook<span class=w>  </span>--ip<span class=o>=</span><span class=m>0</span>.0.0.0<span class=w> </span>--allow-root
</code></pre></div> </div> <h2 id=cleaning-up>Cleaning up<a class=headerlink href=#cleaning-up title="Permanent link">#</a></h2> <p>As mentioned before, Docker tends to consume a lot of disk space. In general, <code>docker image rm</code> is used for removing images and <code>docker container rm</code> for removing containers. Here are some convenient commands for cleaning up.</p> <div class=highlight><pre><span></span><code><span class=c1># Remove unused images</span>
docker<span class=w> </span>image<span class=w> </span>prune

<span class=c1># Remove stopped containers</span>
docker<span class=w> </span>container<span class=w> </span>prune

<span class=c1># Remove unused volumes (not used here, but included for reference)</span>
docker<span class=w> </span>volume<span class=w> </span>prune

<span class=c1># Stop and remove ALL containers</span>
docker<span class=w> </span>container<span class=w> </span>rm<span class=w> </span><span class=k>$(</span>docker<span class=w> </span>container<span class=w> </span>ls<span class=w> </span>-a<span class=w> </span>-q<span class=k>)</span>

<span class=c1># Remove ALL images</span>
docker<span class=w> </span>image<span class=w> </span>rm<span class=w> </span><span class=k>$(</span>docker<span class=w> </span>image<span class=w> </span>ls<span class=w> </span>-a<span class=w> </span>-q<span class=k>)</span>
</code></pre></div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../../../..", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "header.autohide", "navigation.footer", "navigation.indexes", "navigation.tabs", "navigation.top", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "toc.integrate"], "search": "../../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../../../../assets/javascripts/bundle.b4d07000.min.js></script> </body> </html>